{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reads Assignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys, os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from starmap.sequencing import *\n",
    "from scipy.io import loadmat\n",
    "from tqdm import tqdm\n",
    "import tifffile\n",
    "from skimage.measure import regionprops"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import wraps\n",
    "from time import time\n",
    "\n",
    "# Timer\n",
    "def timer(func):\n",
    "    @wraps(func)\n",
    "    def _time_it(*args, **kwargs):\n",
    "        start = int(round(time() * 1000))\n",
    "        try:\n",
    "            return func(*args, **kwargs)\n",
    "        finally:\n",
    "            end_ = int(round(time() * 1000)) - start\n",
    "            end_ = round(end_ / 1000, 4)\n",
    "            print(f\"Total execution time: {end_ if end_ > 0 else 0} s\")\n",
    "    return _time_it\n",
    "\n",
    "@timer\n",
    "# Load reads and their positions from mat file\n",
    "def load_reads(fpath, reads_file):\n",
    "    S = loadmat(os.path.join(fpath, reads_file))\n",
    "    bases = [str(i[0][0]) for i in S[\"merged_reads\"]]\n",
    "    points = S[\"merged_points\"]\n",
    "    temp = np.zeros(points.shape)\n",
    "    temp[:, 0] = np.round(points[:, 2]-1) \n",
    "    temp[:, 1] = np.round(points[:, 1]-1)\n",
    "    temp[:, 2] = np.round(points[:, 0]-1)\n",
    "    \n",
    "    print(f\"Number of reads: {len(bases)}\")\n",
    "    \n",
    "    return bases, temp\n",
    "\n",
    "# Load gene table from genes.csv\n",
    "def load_genes(fpath):\n",
    "    genes2seq = {}\n",
    "    seq2genes = {}\n",
    "    with open(os.path.join(fpath, \"genes.csv\"), encoding='utf-8-sig') as f:\n",
    "        for l in f:\n",
    "            fields = l.rstrip().split(\",\")\n",
    "            genes2seq[fields[0]] = \"\".join([str(s+1) for s in encode_SOLID(fields[1][::-1])])\n",
    "            seq2genes[genes2seq[fields[0]]] = fields[0]\n",
    "    return genes2seq, seq2genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# sequencing data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# IO path \n",
    "ppath = 'Z:/Data/Processed/2022-09-30-Rena-Cardiomyocyte64Gene/'\n",
    "opath = os.path.join(ppath, 'output')\n",
    "areas = range(1,7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load genes\n",
    "genes2seqs, seqs2genes = load_genes(ppath)\n",
    "genes = genes2seqs.keys()\n",
    "genes_to_index = {}  \n",
    "for i, k in enumerate(genes):\n",
    "    genes_to_index[k] = i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reads: 450801\n",
      "Total execution time: 2.968 s\n"
     ]
    }
   ],
   "source": [
    "area_num = 6\n",
    "# Load reads of current well\n",
    "bases, points = load_reads(os.path.join(opath, '2022-10-15-merged-points', f'Area_{area_num}'), \"merged_goodPoints.mat\")\n",
    "bases = np.array([seqs2genes[x] for x in bases])\n",
    "points = np.array(points)\n",
    "points = points.astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "====Processing: whole_cell====\n",
      "Total number of cells: 1389\n",
      "Getting region information...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1389/1389 [01:49<00:00, 12.69it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "82.42 % [371536 out of 450801] reads were assigned to 1389 cells\n",
      "====Processing: nucleus====\n",
      "Getting region information...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1389/1389 [00:17<00:00, 81.21it/s] \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33.84 % [152546 out of 450801] reads were assigned to 1389 cells\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAEFCAYAAAAMk/uQAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAWvUlEQVR4nO3de5Bc5Xnn8e8zI+ELVwcmRFxkgRfIOmQtvBOoGNsQe5PFWeJLyiFQrEMwQXatseM1dmI7VetsqlLl3fUtxAnZMShALRdjYzZk14thHVbEu8a2hAmSECKARSEkkCIwCGO6Z6af/aOPjlpSj6ZnpO4z3f39VE1N93v6zHmOKOY3533f857ITCRJAhipugBJ0sJhKEiSSoaCJKlkKEiSSoaCJKm0qOoCDsQxxxyTy5Ytq7oMSeora9as+afMHGu3ra9DYdmyZaxevbrqMiSpr0TEEzNts/tIklQyFCRJJUNBklQyFCRJJUNBklQyFCRJJUNBklQyFCRJJUOhRWZSq9XwGROShlXXQiEiVkbEtohY19L21Yh4oPjaFBEPFO3LIuKnLdv+qlt17U+9Xue9X7qTnTt3VnF4SapcN5e5uA74MnDDrobM/O1dryPi88DzLZ9/LDOXd7GejowuWlx1CZJUma6FQmbeGxHL2m2LiAAuAN7WreNLkuauqjGFtwDPZOY/trSdFBE/jIhVEfGWmXaMiBURsToiVm/fvr37lUrSEKkqFC4Cbm55vxVYmplnAB8DboqII9rtmJkTmTmemeNjY21XfpUkzVPPQyEiFgG/CXx1V1tm1jJzR/F6DfAYcGqva5OkYVfFlcK/Ah7OzM27GiJiLCJGi9cnA6cAj1dQmyQNtW5OSb0Z+C5wWkRsjojLik0XsmfXEcBbgQeLKapfBz6Ymc92qzZJUnvdnH100Qztv9um7Tbgtm7VIknqTF8/jrOfrPjIx9my4/l92o87+kgmrvpcBRVJ0r4MhRa1Wo1GoztLXGzZ8TyHnfP+fdtXrezK8SRpPlz7SJJUMhQkSSVDQZJUMhQkSSVDQZJUMhQkSSVDQZJUMhQkSSVDQZJUMhQkSSVDQZJUMhQkSSUXxOuCdiuirt+wkbPOqaggSeqQodAF7VZErT14ZUXVSFLn7D6SJJUMBUlSyVCQJJUMBUlSqWuhEBErI2JbRKxrafvjiHgqIh4ovn69ZdunIuLRiNgYEf+6W3VJkmbWzSuF64Dz2rR/MTOXF1/fBIiI1wMXAr9Q7POXETHaxdokSW10LRQy817g2Q4//i7glsysZeaPgEeBM7tVmySpvSrGFK6IiAeL7qXXFG3HA0+2fGZz0baPiFgREasjYvX27du7XaskDZVeh8LVwOuA5cBW4PNz/QGZOZGZ45k5PjY2dpDLk6Th1tNQyMxnMnM6MxvAV9jdRfQUcGLLR08o2iRJPdTTZS4iYklmbi3evgfYNTPpDuCmiPgCcBxwCvD9XtY2H+3WOALXOZLUv7oWChFxM3AucExEbAY+A5wbEcuBBDYBHwDIzPURcSvwEDAFfCgzp7tV28HSbo0jcJ0jSf2ra6GQmRe1ab52P5//U+BPu1WPJGl23tEsSSoZCpKkkqEgSSoZCpKkkqEgSSoZCpKkkqEgSSr19I5m7Wvd2rWcf/Hle7Qdd/SRTFz1uYoqkjTMDIWK1XNkn7uit6xaWVE1koad3UeSpJKhIEkqGQqSpJKhIEkqGQqSpJKhIEkqGQqSpJKhIEkqGQqSpJKhIEkquczFAuR6SJKq0rVQiIiVwPnAtsw8vWj7L8BvAHXgMeDSzPxxRCwDNgAbi93vy8wPdqu2hc71kCRVpZvdR9cB5+3Vdjdwemb+C+AR4FMt2x7LzOXF19AGgiRVqWuhkJn3As/u1XZXZk4Vb+8DTujW8SVJc1flQPP7gf/V8v6kiPhhRKyKiLfMtFNErIiI1RGxevv27d2vUpKGSCWhEBF/BEwBNxZNW4GlmXkG8DHgpog4ot2+mTmRmeOZOT42NtabgiVpSPQ8FCLid2kOQF+cmQmQmbXM3FG8XkNzEPrUXtdWHJ9arUZRmiQNlZ6GQkScB/wB8M7MfKmlfSwiRovXJwOnAI/3srZdGtNTXLbyPur1ehWHl6RKdXNK6s3AucAxEbEZ+AzN2UavAO6OCNg99fStwJ9ExCTQAD6Ymc+2/cE9MLLokKoOLUmV6looZOZFbZqvneGztwG3dasWSVJnXOZCklQyFCRJJUNBklQyFCRJJUNBklQyFCRJJUNBklQyFCRJJUNBklQyFCRJJUNBklQyFCRJJUNBklQyFCRJJUNBklQyFCRJJUNBklQyFCRJJUNBklTqKBQi4uxO2iRJ/a3TK4U/77BtDxGxMiK2RcS6lrafiYi7I+Ifi++vKdojIq6KiEcj4sGIeGOHtUmSDpJF+9sYEb8MvAkYi4iPtWw6Ahjt4OdfB3wZuKGl7ZPAtzPzsxHxyeL9HwLvAE4pvs4Cri6+S5J6ZLYrhUOAw2iGx+EtXy8A753th2fmvcCzezW/C7i+eH098O6W9huy6T7gqIhY0sE5SJIOkv1eKWTmKmBVRFyXmU8cpGMem5lbi9dPA8cWr48Hnmz53OaibWtLGxGxAlgBsHTp0oNUkiQJZgmFFq+IiAlgWes+mfm2Azl4ZmZE5Bz3mQAmAMbHx+e0ryRp/zoNha8BfwVcA0wf4DGfiYglmbm16B7aVrQ/BZzY8rkTijZJUo90GgpTmXn1QTrmHcAlwGeL73/T0n5FRNxCc4D5+ZZuJklSD3QaCn8bEf8OuB2o7WrMzL0HkfcQETcD5wLHRMRm4DM0w+DWiLgMeAK4oPj4N4FfBx4FXgIu7fw0JEkHQ6ehcEnx/RMtbQmcvL+dMvOiGTa9vc1nE/hQh/VIkrqgo1DIzJO6XYj2b93atZx/8eX7tB939JFMXPW5CiqSNIg6CoWI+J127Zl5Q7t2HXz1HOGwc96/T/uWVSsrqEbSoOq0++iXWl6/kmb3z/3seaeyJKnPddp99OHW9xFxFHBLNwrS3LTrVrJLSdJ8dXqlsLefAAM1zpCZ1Go1muPn/aNdt5JdSpLmq9Mxhb9l92/LUeCfA7d2q6gq1Ot1LplYhY+YkDTMOr1SaO2LmAKeyMzNXainUqOLFjM9daA3bEtS/+roz+JiYbyHaa6Q+hqg3s2iqjY9VS+6kiRpuHT65LULgO8Dv0XzDuTvRcSsS2dLkvpLp91HfwT8UmZuA4iIMeB/A1/vVmGSpN7rdFR1ZFcgFHbMYV9JUp/o9Erhzoj4FnBz8f63aS5gJ0kaILM9o/mf0XxS2ici4jeBNxebvgvc2O3iJEm9NduVwpeATwFk5jeAbwBExC8W236ji7VJknpstnGBYzNz7d6NRduyrlQkSarMbKFw1H62veog1iFJWgBmC4XVEbHPIv4R8XvAmu6UJEmqymxjCh8Fbo+Ii9kdAuPAIcB7uliXJKkC+w2FzHwGeFNE/ApwetH8PzPz77pemSSp5zp9nsI9wD1drkWSVLH5Pk9h3iLiNOCrLU0nA/+B5qD25cD2ov3TmekNcpLUQz0PhczcCCwHiIhR4CngduBS4IuZ6SPDJKkiVa9f9Hbgscx8ouI6JElUHwoXsns9JYArIuLBiFgZEa+pqihJGlaVhUJEHAK8E/ha0XQ18DqaXUtbgc/PsN+KiFgdEau3b9/e7iOSpHmq8krhHcD9xbRXMvOZzJzOzAbwFeDMdjtl5kRmjmfm+NjYWA/LlaTBV2UoXERL11FELGnZ9h5gXc8rkqQh1/PZRwARcSjwq8AHWpr/c0QsBxLYtNc2SVIPVBIKmfkT4Oi92t5XRS2SpN2qnn0kSVpADAVJUqmS7qN+tOIjH2fLjuf3aFu/YSNnnVNRQZLUBYZCh7bseJ7Dznn/Hm21B6+sqBpJ6g67jyRJJUNBklSy+2gArVu7lvMv3ucpqhx39JFMXOUitJJmZigMoHqO7DP+AbBl1coKqpHUT+w+kiSVDAVJUslQkCSVDAVJUslQaCMzqdVqZGbVpUhSTxkKbTSmp7hs5X3U6/WqS5GknjIUZjCy6JCqS5CknjMUJEklQ0GSVDIUJEklQ0GSVDIUJEmlyhbEi4hNwE5gGpjKzPGI+Bngq8AyYBNwQWY+V1WNkjRsqr5S+JXMXJ6Z48X7TwLfzsxTgG8X7ysxPVWnVqtVdXhJqsRCWzr7XcC5xevrgf8D/GFVxQyzds+k9nkM0uCrMhQSuCsiEvivmTkBHJuZW4vtTwPHVlbdkGv3TGqfxyANvipD4c2Z+VRE/Cxwd0Q83LoxM7MIjD1ExApgBcDSpUt7U6kkDYnKQiEznyq+b4uI24EzgWciYklmbo2IJcC2NvtNABMA4+Pjrlg3B+0e02mXkKRWlYRCRBwKjGTmzuL1rwF/AtwBXAJ8tvj+N1XUN6jaPabTLiFJraq6UjgWuD0idtVwU2beGRE/AG6NiMuAJ4ALelVQrVaj0fDCQ9JwqyQUMvNx4A1t2ncAb+99RZIkWHhTUtVj7cYZANZv2MhZ51RQkKRKGQpDrt04A0DtwSsrqEZS1aq+o1mStIAYCpKkkqEgSSoZCpKkkqEgSSoZCpKkkqEgSSoZCpKkkqEgSSoZCpKkkqEgSSoZCh3KTKYn62S6vLakwWUodKjRaLB+09M0piarLkWSusZQmINYtLjqEiSpqwwFSVLJ5yl0IDNpTE8z7KMJ7R7Ic9zRRzJx1ecqqkjSwWYodKBer/P4EW9g8ZAPMrd7IM+WVSsrqkZSN9h9NIvMpFarwaj5KWnwGQqzqNfrXPjndzkVVdJQ6PmfvxFxInADcCyQwERm/llE/DFwObC9+OinM/Obva6vnVFnHc2o3TgDONYg9asq+kSmgCsz8/6IOBxYExF3F9u+mJn+Jukj7cYZwLEGqV/1PBQycyuwtXi9MyI2AMf3uo7ZZCYvv/yy3UaShkqlo6cRsQw4A/gecDZwRUT8DrCa5tXEc232WQGsAFi6dGnXamtMT3HpV75DjIyQQE5PGRCSBl5lA80RcRhwG/DRzHwBuBp4HbCc5pXE59vtl5kTmTmemeNjY2NdrXFk0WJGHE+QNEQqCYWIWEwzEG7MzG8AZOYzmTmdmQ3gK8CZVdTWanpqkmw0qi5Dknqm56EQEQFcC2zIzC+0tC9p+dh7gHW9rk2Shl0VYwpnA+8D1kbEA0Xbp4GLImI5zWmqm4APVFCbJA21KmYffQeINpsWxD0JkjTMvKOZlqUsZlnyrjE1yfRkvTdFSVIFXNCH5lIWl0yswoyUNOyGPhR2XSWMLlrM9NR01eUMDJfZlvrT0IdCvV7n4qvvodGY/ca0zGR6qvmc5uYkKs3EZbal/mR/CZ0veJfTU2zc+oLPaZY0sAwF5naTms9pljTIDAVJUslQkCSVhn6gWb0zlwfyrPjIx9my4/mOPivp4DEU9vKjJ56gPrV7fCFGF8H0FC//9Ke8+vAKCxsAc3kgz5Ydz7f97F1f/vdOdZW6yFDYS32qwcjhu5fkHlm0mJyeovGTTdUVpZJTXaXuGvoxhVqt1tE9CpI0DIY+FCRJuxkKc+SieJIGmaEwD5nJ9GTdZzZLGjhDPdD88ssv88ILLzDbktmtMpPJ2ks8vOU5stFgerLOyKLFroW0wMw0pfXxRx7m5FN/fo82Zy9Juw1tKGQmO3fu5PK//n/ESOf/DDk9xSNbX2D0la8mp6dZt+lpTl/2c4wuPqSL1Q62dvcvrN+wkbPOOcD9P/if9vnsjx+80tlL0n4MbSjseobCXAJhlyimqY6e8W4YGfVq4QC1m2Zae/DKnu0/k3ZXG15VaNANbSjUajViZFHHC+G1KldJHRklp6fYsHkHp590iFcLfWqmO63bXW14VaFBN7ShcDC1Wzk1M2lMTe5xBdGuTdWb6U7rg3G1cSBc6kNVWHChEBHnAX8GjALXZOZnKy5pvxpTk82lMArTk3Wy0aAxNcn6TU/zCy3jDe3apJnMtNRHP12t2AXXfxZUKETEKPAXwK8Cm4EfRMQdmflQtZXtX2YyVa/RKMKgUcxKytFF5XjDrmmsdHhVIQ2CdsHWT6E2jBZUKABnAo9m5uMAEXEL8C6gK6HQaPdwnekpsuXJaklzxlHu1Q7QqL9MY7JOY7LO2kd2EKPNX+qNyTrrHnuSGF3MQ0+8xOtf+3Pk9DTrH9/MyCGvov7TFxmdbI5BZKPByzufY+NTz/L61zavIBpTU9RfenGPq4ld01/LMifrXm1UYKbxh3ZTXTttm6l9LjOwZtLuL/W5TMvtdP/9/Qy11+m/ba//vWMh3YAVEe8FzsvM3yvevw84KzOvaPnMCmBF8fY0YOM8D3cM8E8HUO5CMQjnMQjnAINxHoNwDjAY59HNc3htZo6127DQrhRmlZkTwMSB/pyIWJ2Z4wehpEoNwnkMwjnAYJzHIJwDDMZ5VHUOC22Zi6eAE1ven1C0SZJ6YKGFwg+AUyLipIg4BLgQuKPimiRpaCyo7qPMnIqIK4Bv0ZySujIz13fpcAfcBbVADMJ5DMI5wGCcxyCcAwzGeVRyDgtqoFmSVK2F1n0kSaqQoSBJKg1lKETEeRGxMSIejYhPVl3PfETEyojYFhHrqq5lviLixIi4JyIeioj1EfH7Vdc0VxHxyoj4fkT8Q3EO/7HqmuYrIkYj4ocR8T+qrmW+ImJTRKyNiAciYnXV9cxXRBwVEV+PiIcjYkNE/HLPjj1sYwrFUhqP0LKUBnDRQl9KY28R8VbgReCGzDy96nrmIyKWAEsy8/6IOBxYA7y7n/5bRHNdkkMz88WIWAx8B/j9zLyv4tLmLCI+BowDR2Tm+VXXMx8RsQkYz8y+vnEtIq4H/j4zrylmYr46M3/ci2MP45VCuZRGZtaBXUtp9JXMvBd4tuo6DkRmbs3M+4vXO4ENwPHVVjU32fRi8XZx8dV3f2lFxAnAvwGuqbqWYRcRRwJvBa4FyMx6rwIBhjMUjgeebHm/mT77RTSIImIZcAbwvYpLmbOi2+UBYBtwd2b23TkAXwL+AJj7A0YWlgTuiog1xZI4/egkYDvw10V33jURcWivDj6MoaAFJiIOA24DPpqZL1Rdz1xl5nRmLqd5B/6ZEdFX3XkRcT6wLTPXVF3LQfDmzHwj8A7gQ0U3a79ZBLwRuDozzwB+AvRs7HMYQ8GlNBaQoh/+NuDGzPxG1fUciOIS/x7gvIpLmauzgXcW/fG3AG+LiP9WbUnzk5lPFd+3AbfT7C7uN5uBzS1XnF+nGRI9MYyh4FIaC0QxSHstsCEzv1B1PfMREWMRcVTx+lU0JzA8XGlRc5SZn8rMEzJzGc3/H/4uM/9txWXNWUQcWkxYoOhu+TWg72bnZebTwJMRcVrR9Ha69PiAdhbUMhe90OOlNLomIm4GzgWOiYjNwGcy89pqq5qzs4H3AWuLPnmAT2fmN6srac6WANcXs9pGgFszs2+ndPa5Y4HbiwdVLQJuysw7qy1p3j4M3Fj84fo4cGmvDjx0U1IlSTMbxu4jSdIMDAVJUslQkCSVDAVJUslQkKQ+MdeFMCPigpYFJ2/qaB9nH0lSf5jLQpgRcQpwK/C2zHwuIn62uKlvv7xSkKQ+0W4hzIh4XUTcWaz39PcR8fPFpsuBv8jM54p9Zw0EMBQkqd9NAB/OzH8JfBz4y6L9VODUiPi/EXFfRHS0/MrQ3dEsSYOiWEzyTcDXiju5AV5RfF8EnEJz5YMTgHsj4hdnW4bbUJCk/jUC/LhYpXdvm4HvZeYk8KOIeIRmSPxgth8oSepDxVLzP4qI34LmIpMR8YZi83+neZVARBxDszvp8dl+pqEgSX2iWAjzu8BpEbE5Ii4DLgYui4h/ANaz+0mS3wJ2RMRDNJd0/0Rm7pj1GE5JlSTt4pWCJKlkKEiSSoaCJKlkKEiSSoaCJKlkKEiSSoaCJKn0/wE65/NkWcuUkwAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "structure_dict = {'whole_cell': 'cell',\n",
    "                 'nucleus': 'nuclei'}\n",
    "\n",
    "total_cells = None # NOTE only defined in whole_cell \n",
    "# expr_dict = {}\n",
    "\n",
    "for current_structure in structure_dict.keys():\n",
    "    # current_structure = 'nucleus'\n",
    "    print(f\"====Processing: {current_structure}====\")\n",
    "\n",
    "    current_out_path = os.path.join(opath, f'Area{area_num}', current_structure)\n",
    "    if not os.path.exists(current_out_path):\n",
    "        os.mkdir(current_out_path) \n",
    "    \n",
    "    # Load segmentation\n",
    "    seg_path = os.path.join(opath, '2022-10-17-3D-segmentation')\n",
    "    current_seg = tifffile.imread(os.path.join(seg_path, structure_dict[current_structure], f'Area{area_num}_{structure_dict[current_structure]}_3Dseg.tif'))\n",
    "\n",
    "    # remove reads that fall out of frame\n",
    "    in_frame = []\n",
    "    for i, p in enumerate(points):\n",
    "        if np.any(p >= current_seg.shape):\n",
    "            in_frame.append(False)\n",
    "        else: in_frame.append(True)\n",
    "    points = points[in_frame]\n",
    "    bases = bases[in_frame]\n",
    "\n",
    "    # Assign reads to current structure\n",
    "    reads_assignment = current_seg[points[:, 0], points[:, 1], points[:, 2]]\n",
    "\n",
    "    if not total_cells:\n",
    "        total_cells = len(np.unique(current_seg)) - 1\n",
    "        print(f\"Total number of cells: {total_cells}\")\n",
    "\n",
    "    cell_by_barcode = np.zeros((total_cells, len(genes)))\n",
    "\n",
    "    # Get region information \n",
    "    print('Getting region information...') \n",
    "    rp = regionprops(current_seg)\n",
    "\n",
    "    region_labels = []\n",
    "    areas_dict = []\n",
    "    cell_locs_dict = []\n",
    "    for i, region in enumerate(tqdm(rp)):\n",
    "        region_labels.append(region.label)\n",
    "        area = region.area\n",
    "        areas_dict.append(area)\n",
    "        if area > 0:\n",
    "            cell_locs_dict.append(region.centroid)\n",
    "        else: cell_locs_dict.append((0,0,0)) # NOTE to be deleted\n",
    "\n",
    "        assigned_reads = bases[reads_assignment == i+1]  # i+1 == region label\n",
    "        for j in assigned_reads:\n",
    "            if j in genes_to_index:\n",
    "                cell_by_barcode[i, genes_to_index[j]] += 1 \n",
    "\n",
    "    cell_by_barcode = cell_by_barcode.astype(int)\n",
    "    cell_locs_dict = np.array(cell_locs_dict).astype(int) #?\n",
    "    curr_meta = pd.DataFrame({'sample': area_num, 'area': areas_dict,\n",
    "                                'x':cell_locs_dict[:, 1], 'y':cell_locs_dict[:, 2], 'z':cell_locs_dict[:, 0]})\n",
    "\n",
    "    # Output\n",
    "    with open(os.path.join(current_out_path, \"log.txt\"), 'w') as f:\n",
    "        msg = f'{100*cell_by_barcode.sum()/len(bases):.2f} % [{cell_by_barcode.sum()} out of {len(bases)}] reads were assigned to {total_cells} cells'\n",
    "        print(msg)\n",
    "        f.write(msg)\n",
    "    np.savetxt(os.path.join(current_out_path, \"cell_barcode_count.csv\"), cell_by_barcode, delimiter=',', fmt=\"%d\")\n",
    "    cell_barcode_names = pd.DataFrame({'seq': list(seqs2genes.keys()), 'gene': list(seqs2genes.values())})\n",
    "    cell_barcode_names.to_csv(os.path.join(current_out_path, \"cell_barcode_names.csv\"), header=False)\n",
    "    curr_meta.to_csv(os.path.join(current_out_path, \"meta.csv\"))\n",
    "\n",
    "    # expr_dict[current_structure] = cell_by_barcode\n",
    "    # print(curr_meta.head())\n",
    "    sns.histplot(areas_dict)\n",
    "    # break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate complete matrix for all samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "====Generating: whole_cell====\n",
      "====Generating: nucleus====\n"
     ]
    }
   ],
   "source": [
    "# Construct complete matrix # NOTE just to concat samples together\n",
    "for current_structure in structure_dict.keys():\n",
    "    print(f\"====Generating: {current_structure}====\")\n",
    "    cell_by_gene_complete = None \n",
    "    meta_complete = None    \n",
    "\n",
    "    for i, area_num in enumerate(areas):\n",
    "        current_expr_path = os.path.join(opath, f'Area{area_num}')\n",
    "        current_expr = np.loadtxt(os.path.join(current_expr_path, current_structure, \"cell_barcode_count.csv\"), dtype=int, delimiter=',')\n",
    "        current_meta = pd.read_csv(os.path.join(current_expr_path, current_structure, \"meta.csv\")) \n",
    "\n",
    "        # add to complete matrix\n",
    "        if cell_by_gene_complete is not None:\n",
    "            cell_by_gene_complete = np.concatenate((cell_by_gene_complete, current_expr))\n",
    "        else:\n",
    "            cell_by_gene_complete = current_expr\n",
    "\n",
    "        if meta_complete is not None:\n",
    "            meta_complete = pd.concat([meta_complete, current_meta])\n",
    "        else:\n",
    "            meta_complete = current_meta\n",
    "\n",
    "    current_output_path = os.path.join(opath, current_structure)\n",
    "    if not os.path.exists(current_output_path):\n",
    "        os.mkdir(current_output_path)\n",
    "    np.savetxt(os.path.join(current_output_path, \"complete_cell_barcode_count.csv\"), cell_by_gene_complete.astype(int), delimiter=',', fmt=\"%d\")\n",
    "    meta_complete = meta_complete.reset_index(drop=True)\n",
    "    meta_complete = meta_complete.rename(columns={\"Unnamed: 0\": \"orig_index\"})\n",
    "    meta_complete.to_csv(os.path.join(current_output_path, \"complete_meta.csv\"))\n",
    "    # cell_barcode_names.to_csv(os.path.join(current_output_path, \"cell_barcode_names.csv\"), header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "0774c9394a1aea0cca30f13f7f373026b44b800bfe001e3b93c40d9a34a50f91"
  },
  "kernelspec": {
   "display_name": "Python 3.8.11 ('connie')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
