{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reads Assignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys, os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from starmap.sequencing import *\n",
    "from scipy.io import loadmat\n",
    "from tqdm import tqdm\n",
    "import tifffile\n",
    "from skimage.measure import regionprops"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import wraps\n",
    "from time import time\n",
    "\n",
    "# Timer\n",
    "def timer(func):\n",
    "    @wraps(func)\n",
    "    def _time_it(*args, **kwargs):\n",
    "        start = int(round(time() * 1000))\n",
    "        try:\n",
    "            return func(*args, **kwargs)\n",
    "        finally:\n",
    "            end_ = int(round(time() * 1000)) - start\n",
    "            end_ = round(end_ / 1000, 4)\n",
    "            print(f\"Total execution time: {end_ if end_ > 0 else 0} s\")\n",
    "    return _time_it\n",
    "\n",
    "@timer\n",
    "# Load reads and their positions from mat file\n",
    "def load_reads(fpath, reads_file):\n",
    "    S = loadmat(os.path.join(fpath, reads_file))\n",
    "    bases = [str(i[0][0]) for i in S[\"merged_reads\"]]\n",
    "    points = S[\"merged_points\"]\n",
    "    temp = np.zeros(points.shape)\n",
    "    temp[:, 0] = np.round(points[:, 2]-1) \n",
    "    temp[:, 1] = np.round(points[:, 1]-1)\n",
    "    temp[:, 2] = np.round(points[:, 0]-1)\n",
    "    \n",
    "    print(f\"Number of reads: {len(bases)}\")\n",
    "    \n",
    "    return bases, temp\n",
    "\n",
    "# Load gene table from genes.csv\n",
    "def load_genes(fpath):\n",
    "    genes2seq = {}\n",
    "    seq2genes = {}\n",
    "    with open(os.path.join(fpath, \"genes.csv\"), encoding='utf-8-sig') as f:\n",
    "        for l in f:\n",
    "            fields = l.rstrip().split(\",\")\n",
    "            genes2seq[fields[0]] = \"\".join([str(s+1) for s in encode_SOLID(fields[1][::-1])])\n",
    "            seq2genes[genes2seq[fields[0]]] = fields[0]\n",
    "    return genes2seq, seq2genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# sequencing data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# IO path \n",
    "ppath = 'Z:/Data/Processed/2022-09-12-Rena-HeLa16Gene/'\n",
    "opath = os.path.join(ppath, 'output')\n",
    "areas = range(1,8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load genes\n",
    "genes2seqs, seqs2genes = load_genes(ppath)\n",
    "genes = genes2seqs.keys()\n",
    "genes_to_index = {}  \n",
    "for i, k in enumerate(genes):\n",
    "    genes_to_index[k] = i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reads: 665859\n",
      "Total execution time: 4.49 s\n"
     ]
    }
   ],
   "source": [
    "area_num = 7\n",
    "# Load reads of current well\n",
    "bases, points = load_reads(os.path.join(opath, '2022-09-22-merged-points', f'Area_{area_num}'), \"merged_goodPoints.mat\")\n",
    "bases = np.array([seqs2genes[x] for x in bases])\n",
    "points = np.array(points)\n",
    "points = points.astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "====Processing: whole_cell====\n",
      "Total number of cells: 512\n",
      "Getting region information...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 512/512 [00:55<00:00,  9.31it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "86.60 % [576655 out of 665859] reads were assigned to 512 cells\n",
      "====Processing: nucleus====\n",
      "Getting region information...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 512/512 [00:13<00:00, 37.64it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26.49 % [176387 out of 665859] reads were assigned to 512 cells\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYMAAAEFCAYAAAABjYvXAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAUsUlEQVR4nO3dfZBddX3H8fc3WcJTgC15QNjdEIxIqrG2GCkFx2mhnWGqFdtSYIZabFG04lNtbbWdqdPOOGNnOq3U0mIKtmitBpHWSK1Oq1TbYlMX1CpwY2gg7OaBXTAJEB7CzX77xz0bNssmeza75557s+/XzA734dzcT06497PnnN/5nchMJEnz24K6A0iS6mcZSJIsA0mSZSBJwjKQJAE9dQcoY+nSpbly5cq6Y0hSV7n77rsfzcxlZZbtijJYuXIlg4ODdceQpK4SEVvLLutuIkmSZSBJsgwkSVgGkiQsA0kSloEkCctAkoRlIEmiS0466yTNZpNGo3Hg/urVq+npcTVK6m5+i81Qo9Hg7TfcweLl/Tw5MsyN18GaNWvqjiVJs2IZHIHFy/vp7VtVdwxJmjMeM5AkWQaSJMtAkoRlIEnCMpAkYRlIkrAMJElYBpIkLANJEpaBJAnLQJKEZSBJwjKQJGEZSJKwDCRJWAaSJCwDSRKWgSQJy0CShGUgScIykCRhGUiSsAwkSVgGkiQsA0kSloEkCctAkkTFZRARvxUR90bE9yPiMxFxXEScFREbI+KBiFgfEYuqzCBJml5lZRARfcC7gbWZuQZYCFwJ/Anw55n5EmAXcE1VGSRJ5VS9m6gHOD4ieoATgB3ARcBtxfO3AG+sOIMkaRqVlUFmbgP+FHiYVgnsAe4Gdmdms1hsGOib6vURcW1EDEbE4OjoaFUxJUlUu5voR4BLgbOAM4ATgUvKvj4z12Xm2sxcu2zZsopSSpKg2t1EPws8mJmjmfkccDtwIdBb7DYC6Ae2VZhBklRClWXwMHB+RJwQEQFcDNwH3AlcVixzNfCFCjNIkkqo8pjBRloHiu8Bvle81zrg94D3RcQDwBLg5qoySJLK6Zl+kSOXmR8CPjTp4S3AeVW+ryRpZjwDWZJkGUiSLANJEpaBJAnLQJKEZSBJwjKQJGEZSJKwDCRJWAaSJCwDSRKWgSQJy0DT6BtYQUSU/ukbWFF3ZElHoNJZS9X9tg8PccXH7yq9/Pq3XVBhGklVcctAkmQZSJIsA0kSloEkCctAc21BjyOPpC7kaCLNrbFm6dFHjjySOodbBpIky0CSZBlIkrAMJElYBpIkLANJEpaBJAnLQJKEJ51Nq9ls0mg0DtzfvHkzZI2BJKkClsE0Go0Gb7/hDhYv7wfgkcYgJ5/5cnrrjSVJc8oyKGHx8n56+1YB8MTIUM1pJGnuecxAkmQZSJIsA0kSlsG81DewovQ1ByTND5UeQI6IXuAmYA2tAZm/AWwC1gMrgYeAyzNzV5U5dLDtw0Nec0DSQareMrge+HJmrgZeCdwPfAD4amaeDXy1uC9JqlFlZRARpwCvBW4GyMx9mbkbuBS4pVjsFuCNVWWQJJVT5ZbBWcAo8LcR8e2IuCkiTgROy8wdxTI7gdOmenFEXBsRgxExODo6WmFMSVKVxwx6gHOBd2Xmxoi4nkm7hDIzI2LKyR0ycx2wDmDt2rUdOQHE2Nj+1vQUE6xevZqeHs/lk9RdqvzWGgaGM3Njcf82WmXwSEScnpk7IuJ0YKTCDJXa++h2PrzhQZau2AvAkyPD3HgdrFmzpuZkkjQzlZVBZu6MiKGIOCczNwEXA/cVP1cDHyn++4WqMrTDicv6DkxVIUndqur9Ge8CPh0Ri4AtwK/TOk5xa0RcA2wFLq84gyRpGpWWQWZ+B1g7xVMXV/m+kqSZ8QxkSZJlIEmyDCRJWAaSJEqWQURcWOYxSVJ3Krtl8LGSj0mSutBhh5ZGxE8BFwDLIuJ9E546GVhYZTBJUvtMd57BImBxsdxJEx5/HLisqlCSpPY6bBlk5teBr0fE32Xm1jZlkiS1WdkzkI+NiHW0rk524DWZeVEVoSRJ7VW2DD4H3EjrEpb7q4sjSapD2TJoZuZfV5pEklSbskNLvxgR74iI0yPi1PGfSpNJktqm7JbB1cV/3z/hsQRePLdxJEl1KFUGmXlW1UE0O30DK9g+PFR3DEldqlQZRMSvTfV4Zn5ybuPoSG0fHuKKj99Vatn1b7ug4jSSuk3Z3USvnnD7OFoXp7kHsAx05Bb0EBGlFj2jf4BtQw9XHEiav8ruJnrXxPsR0Qt8topAmkfGmm7NSB3iSKew3gt4HEGSjhJljxl8kdboIWhNUPejwK1VhZIktVfZYwZ/OuF2E9iamcMV5JEk1aDUbqJiwroGrZlLfwTYV2UoSVJ7lb3S2eXA/wC/AlwObIwIp7CWpKNE2d1EfwC8OjNHACJiGfBvwG1VBZMktU/ZMlgwXgSFxzjykUhHrbGx/WzevPnA/dWrV9PTU3YVS1J9yn5TfTkivgJ8prh/BfClaiJ1r72PbufDGx5k6Yq9PDkyzI3XwZo1a+qOJUnTmu4ayC8BTsvM90fELwGvKZ76JvDpqsN1oxOX9dHbt6ruGJI0I9NtGXwU+CBAZt4O3A4QEa8onvuFCrNJktpkuv3+p2Xm9yY/WDy2spJEkqS2m64Meg/z3PFzmEOSVKPpdhMNRsRbM/NvJj4YEW8B7q4uVvdzZJGkbjLdt9N7gX+MiKt4/st/LbAI+MUKc3U9RxZJ6iaHLYPMfAS4ICJ+Bhj/JvvnzPxa5cmOAo4sktQtyl7P4E7gzoqzSJJqUvlZxBGxMCK+HRF3FPfPioiNEfFARKyPiEVVZ5AkHV47ppR4D3D/hPt/Avx5Zr4E2AVc04YMkqTDqLQMIqIfeB1wU3E/gIt4foK7W4A3VplBkjS9qrcMPgr8LjBW3F8C7M7MZnF/GOirOIMkaRqVlUFEvB4YycwjOh8hIq6NiMGIGBwdHZ3jdJKkiarcMrgQeENEPAR8ltbuoeuB3ogYH8XUD2yb6sWZuS4z12bm2mXLllUYU5JUWRlk5gczsz8zVwJXAl/LzKtoDVEdv0ra1cAXqsogSSqnjgvU/B7wvoh4gNYxhJtryCBJmqAtk+Vk5r8D/17c3gKc1473lSSV46UrJUmWQSfrG1hBRJT6kaTZcE7lDrZ9eIgrPn5XqWXXv+2CitNIOpq5ZSBJsgwkSZaBJAnLQN1iQU/pg+l9AyvqTit1HQ8gqzuMNT2YLlXILQNJkmUgSbIMJElYBpIkLANJEpaBJAnLQEcjz0mQZszzDNpgbGw/mzdvPuix1atX09Pj6q+E5yRIM+a3URvsfXQ7H97wIEtX7AXgyZFhbrwO1qxZU3MySWqxDNrkxGV99PatqjuGJE3JMqjB5N1G7jKSVDe/gWowcbeRu4wkdQLLoCbuNpLUSRxaKklyy2AqzWaTRqMB0Nq3nzUHkqSKWQZTaDQavP2GO1i8vJ9HGoOcfObL6a07lCRVyN1Eh7B4eT+9fas44dTT6o4iSZWzDCRJ7iaqm1NVSOoEfuPUzKkqJHUCy6ADeM6BpLp5zECSZBlIkiwDSRIeM2iru+76JvuefZZnhu6l56QlbBp69KDbAM3dO3nFH7+p5qSS5pt5WwYTp5xoNpsAB4ZzVjUFxb5nn2X5OeeyZ+xpFp60hMX9qw+6DfDMyEOs/cNP0du3yqtwSWqbeVsGk6ecWHhCL0tXvATAKSjmk+J6yWWd0T/AtqGHKwwk1aOyMoiIAeCTwGm0fs9el5nXR8SpwHpgJfAQcHlm7qoqx+GMTznxxMgQPSctPTC884mRoTriqA4zuF4yeM1kHb2qPIDcBH47M18GnA9cFxEvAz4AfDUzzwa+WtyXJNWosjLIzB2ZeU9x+wngfqAPuBS4pVjsFuCNVWWQJJXTlqGlEbES+AlgI3BaZu4ontpJazfSVK+5NiIGI2JwdHS0HTElad6q/AByRCwGPg+8NzMfn3iwLjMzIqYct5OZ64B1AGvXrp312J6Jo4fAi9ZI0kSVlkFEHEOrCD6dmbcXDz8SEadn5o6IOB0YqTLDuImjh8ARQ5I0UWW7iaK1CXAzcH9m/tmEpzYAVxe3rwa+UFWGycZHD3nRGkk6WJVbBhcCbwK+FxHfKR77feAjwK0RcQ2wFbi8wgySpBIqK4PM/E/gUGfzXFzV+0qSZs6J6iRJloEkyTKQJGEZSDNTTGxX5qdvYEXdaaXS5u2spdIRmcHEdk5qp27iloEkyS2DTpNj+w+aQntsf5MFC/1nklQttww6zL5dO9n6yB62jO7llNdcxeM7t9YdSdI8YBl0oEWnns5xy1fSs3hJ3VEkzROWgSTJMpAkWQaSJBxNNGubGpvYNPQoAM8M3UvPSUvYNPTolLd7el9Uc1pJmpplMEvN557jjHPOBWDP2NMsPGkJi/tXT3n7qb1P1ZxWkqbmbiJJkmUgSbIMJElYBpIkLANJEpaBJAnLQJKEZSBJwjKQJOEZyFPa1NgEx+18wbQSMPWUE5LU7SyDKTSfa3Lqqhe/YFoJeOGUE3sf31NZjhwbO+iqZwAnv+hMr3wmac65m6iD7d+768BVz7aM7uX+xg+88tlRqm9gBRFR6qdvYEXdcXUU8lfMDjd+1TMd3bYPD3HFx+8qtez6t11QcRrNR24ZSN1mQY9bEZpzbhlI3Was6VaE5pxlIFWl+A1e6gZHdRk0m00ajQYAmzdvhqw50Czl2P6DRhc5sqjD+Ru8ushR/U3SaDR4+w13sHh5P480Bjn5zJfTW3eoWdi3aydbn9vHCXEqzz62jR8FevtW1R1L0lHgqC4DgMXL++ntW/WC8frdytFFkqrgaCJJB3TC+Q6dkGE+Ouq3DCSV1wnnO3RChvmoljKIiEuA64GFwE2Z+ZE6cnSzyQeTAcb2N19wQHlsf/Ogs5bHxvYDsHvb/wGHPgg98XXTLTvZ+PIzeY2kerX9UxoRC4EbgJ8DhoFvRcSGzLyv3Vm62cSDyQCnvOYqHt+59QUHlB/fuZX7Gz/g2CV9ADyx5duc8pqr2DK697AHocdft+SSd0+77GTj7+dB7g5Q5fDWGfzZC485lv3PPVtNjor0Daxg+3C5Y40z+fvNZNkz+gfYNvRwqWVnq45f2c4DHsjMLQAR8VngUqCSMnhyZBiAp374CAufeZbdxx//gvuTn9v/1G72/XAHzyw6jn17Rlj43D6eWXQcwEH39+0Zobn3SZ4ZeWjK5ybfbj79NM+MPHTI5Sb/GRNzTPnnn9B70N91qoPk0x04P9TzR/JnqQPNYHgrzHC3ywyHznbbrp+Z7q7q9nURme0dfB8RlwGXZOZbivtvAn4yM985ablrgWuLu+cAm0q+xVLg0TmKW5VOz2i+2THf7JhvdibmOzMzl5V5UcfuzM3MdcC6mb4uIgYzc20FkeZMp2c03+yYb3bMNztHmq+OoaXbgIEJ9/uLxyRJNamjDL4FnB0RZ0XEIuBKYEMNOSRJhbbvJsrMZkS8E/gKraGln8jMe+fwLWa8a6kGnZ7RfLNjvtkx3+wcUb62H0CWJHUep6OQJFkGkqQuLoOIuCQiNkXEAxHxgSmePzYi1hfPb4yIlR2W780RMRoR3yl+3tLmfJ+IiJGI+P4hno+I+Isi//9GxLkdlu+nI2LPhPX3h23MNhARd0bEfRFxb0S8Z4plalt/JfPVtv6K9z8uIv4nIr5bZPyjKZap7TNcMl/dn+GFEfHtiLhjiudmvu4ys+t+aB14/j/gxcAi4LvAyyYt8w7gxuL2lcD6Dsv3ZuAva1yHrwXOBb5/iOd/HvgXIIDzgY0dlu+ngTtqWnenA+cWt08CfjDFv29t669kvtrWX/H+ASwubh8DbATOn7RMnZ/hMvnq/gy/D/iHqf4dj2TddeuWwYEpLTJzHzA+pcVElwK3FLdvAy6OyiZpOaJ8tcrMbwA/PMwilwKfzJb/Bnoj4vT2pCuVrzaZuSMz7yluPwHcD/RNWqy29VcyX62K9fJkcfeY4mfyaJbaPsMl89UmIvqB1wE3HWKRGa+7bi2DPmDiRDnDvPB/9gPLZGYT2AMsaUu6cvkAfrnYhXBbRAxM8Xydyv4d6vRTxWb8v0TEy+sIUGx+/wSt3xwn6oj1d5h8UPP6K3ZzfAcYAf41Mw+5Dmv4DJfJB/V9hj8K/C4wdojnZ7zuurUMjgZfBFZm5o8B/8rzLa5y7qE178orgY8B/9TuABGxGPg88N7MfLzd7z+dafLVvv4yc39m/jitWQjOi4g17c5wOCXy1fIZjojXAyOZefdc/rndWgZlprQ4sExE9ACnAI+1JV2JfJn5WGaOz2N7E/CqNmUrq6OnDcnMx8c34zPzS8AxEbG0Xe8fEcfQ+qL9dGbePsUita6/6fLVvf4mZdkN3AlcMumpOj/DBxwqX42f4QuBN0TEQ7R2QV8UEX8/aZkZr7tuLYMyU1psAK4ubl8GfC2LoymdkG/S/uM30Nqv20k2AL9WjIo5H9iTmTvqDjUuIl40vg80Is6j9f9yW74oive9Gbg/M//sEIvVtv7K5Ktz/RXvuSwieovbx9O6vklj0mK1fYbL5KvrM5yZH8zM/sxcSeu75WuZ+auTFpvxuuvYWUsPJw8xpUVE/DEwmJkbaH0YPhURD9A6EHllh+V7d0S8AWgW+d7crnwAEfEZWiNKlkbEMPAhWgfJyMwbgS/RGhHzAPAU8Osdlu8y4Dcjogk8DVzZxrK/EHgT8L1inzLA7wMrJuSrc/2VyVfn+oPWiKdbonWxqwXArZl5R6d8hkvmq/UzPNls153TUUiSunY3kSRpDlkGkiTLQJJkGUiSsAwkqePENBM1TrH85fH8xIT/cETv6WgiSeosEfFa4Ela81sd9szsiDgbuBW4KDN3RcTyzByZ6Xu6ZSBJHWaqiRojYlVEfDki7o6I/4iI1cVTbwVuyMxdxWtnXARgGUhSt1gHvCszXwX8DvBXxeMvBV4aEf8VEf8dEZOn9SilK89AlqT5pJh08ALgcxNmoj62+G8PcDatM/b7gW9ExCuKOZVKswwkqfMtAHYXs6hONkzr4knPAQ9GxA9olcO3ZvoGkqQOVkxB/mBE/AocuKzqK4un/4nWVgHFzLMvBbbM9D0sA0nqMMVEjd8EzomI4Yi4BrgKuCYivgvcy/NXT/wK8FhE3Edrqu33Z+aMZ6B1aKkkyS0DSZJlIEnCMpAkYRlIkrAMJElYBpIkLANJEvD/ZkQrE2h02XQAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "structure_dict = {'whole_cell': 'cell.tif',\n",
    "                 'nucleus': 'nuclei.tif'}\n",
    "\n",
    "total_cells = None # NOTE only defined in whole_cell \n",
    "# expr_dict = {}\n",
    "\n",
    "for current_structure in structure_dict.keys():\n",
    "    # current_structure = 'nucleus'\n",
    "    print(f\"====Processing: {current_structure}====\")\n",
    "\n",
    "    current_out_path = os.path.join(opath, f'Area_{area_num}', current_structure)\n",
    "    if not os.path.exists(current_out_path):\n",
    "        os.mkdir(current_out_path) \n",
    "    \n",
    "    # Load segmentation\n",
    "    seg_path = os.path.join(opath, '2022-09-28-3D-segmentation', f'Area_{area_num}')\n",
    "    current_seg = tifffile.imread(os.path.join(seg_path, structure_dict[current_structure]))\n",
    "\n",
    "    # remove reads that fall out of frame\n",
    "    in_frame = []\n",
    "    for i, p in enumerate(points):\n",
    "        if np.any(p >= current_seg.shape):\n",
    "            in_frame.append(False)\n",
    "        else: in_frame.append(True)\n",
    "    points = points[in_frame]\n",
    "    bases = bases[in_frame]\n",
    "\n",
    "    # Assign reads to current structure\n",
    "    reads_assignment = current_seg[points[:, 0], points[:, 1], points[:, 2]]\n",
    "\n",
    "    if not total_cells:\n",
    "        total_cells = len(np.unique(current_seg)) - 1\n",
    "        print(f\"Total number of cells: {total_cells}\")\n",
    "\n",
    "    cell_by_barcode = np.zeros((total_cells, len(genes)))\n",
    "\n",
    "    # Get region information \n",
    "    print('Getting region information...') \n",
    "    rp = regionprops(current_seg)\n",
    "\n",
    "    region_labels = []\n",
    "    areas_dict = []\n",
    "    cell_locs_dict = []\n",
    "    for i, region in enumerate(tqdm(rp)):\n",
    "        region_labels.append(region.label)\n",
    "        area = region.area\n",
    "        areas_dict.append(area)\n",
    "        if area > 0:\n",
    "            cell_locs_dict.append(region.centroid)\n",
    "        else: cell_locs_dict.append((0,0,0)) # NOTE to be deleted\n",
    "\n",
    "        assigned_reads = bases[reads_assignment == i+1]  # i+1 == region label\n",
    "        for j in assigned_reads:\n",
    "            if j in genes_to_index:\n",
    "                cell_by_barcode[i, genes_to_index[j]] += 1 \n",
    "\n",
    "    cell_by_barcode = cell_by_barcode.astype('int')\n",
    "    cell_locs_dict = np.array(cell_locs_dict).astype(int) #?\n",
    "    curr_meta = pd.DataFrame({'sample': area_num, 'area': areas_dict,\n",
    "                                'x':cell_locs_dict[:, 1], 'y':cell_locs_dict[:, 2], 'z':cell_locs_dict[:, 0]})\n",
    "\n",
    "    # Output\n",
    "    with open(os.path.join(current_out_path, \"log.txt\"), 'w') as f:\n",
    "        msg = f'{100*cell_by_barcode.sum()/len(bases):.2f} % [{cell_by_barcode.sum()} out of {len(bases)}] reads were assigned to {total_cells} cells'\n",
    "        print(msg)\n",
    "        f.write(msg)\n",
    "    np.savetxt(os.path.join(current_out_path, \"cell_barcode_count.csv\"), cell_by_barcode, delimiter=',', fmt=\"%d\")\n",
    "    cell_barcode_names = pd.DataFrame({'seq': list(seqs2genes.keys()), 'gene': list(seqs2genes.values())})\n",
    "    cell_barcode_names.to_csv(os.path.join(current_out_path, \"cell_barcode_names.csv\"), header=False)\n",
    "    curr_meta.to_csv(os.path.join(current_out_path, \"meta.csv\"))\n",
    "\n",
    "    # expr_dict[current_structure] = cell_by_barcode\n",
    "    # print(curr_meta.head())\n",
    "    sns.histplot(areas_dict)\n",
    "    # break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate complete matrix for all samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "====Generating: whole_cell====\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\wangk\\AppData\\Local\\Temp\\ipykernel_19928\\2860121050.py:26: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.\n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  np.savetxt(os.path.join(current_output_path, \"complete_cell_barcode_count.csv\"), cell_by_gene_complete.astype(np.int), delimiter=',', fmt=\"%d\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "====Generating: nucleus====\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\wangk\\AppData\\Local\\Temp\\ipykernel_19928\\2860121050.py:26: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.\n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  np.savetxt(os.path.join(current_output_path, \"complete_cell_barcode_count.csv\"), cell_by_gene_complete.astype(np.int), delimiter=',', fmt=\"%d\")\n"
     ]
    }
   ],
   "source": [
    "# Construct complete matrix # NOTE just to concat samples together\n",
    "for current_structure in structure_dict.keys():\n",
    "    print(f\"====Generating: {current_structure}====\")\n",
    "    cell_by_gene_complete = None \n",
    "    meta_complete = None    \n",
    "\n",
    "    for i, area_num in enumerate(areas):\n",
    "        current_expr_path = os.path.join(opath, f'Area_{area_num}')\n",
    "        current_expr = np.loadtxt(os.path.join(current_expr_path, current_structure, \"cell_barcode_count.csv\"), dtype=int, delimiter=',')\n",
    "        current_meta = pd.read_csv(os.path.join(current_expr_path, current_structure, \"meta.csv\")) \n",
    "\n",
    "        # add to complete matrix\n",
    "        if cell_by_gene_complete is not None:\n",
    "            cell_by_gene_complete = np.concatenate((cell_by_gene_complete, current_expr))\n",
    "        else:\n",
    "            cell_by_gene_complete = current_expr\n",
    "\n",
    "        if meta_complete is not None:\n",
    "            meta_complete = pd.concat([meta_complete, current_meta])\n",
    "        else:\n",
    "            meta_complete = current_meta\n",
    "\n",
    "    current_output_path = os.path.join(opath, current_structure)\n",
    "    if not os.path.exists(current_output_path):\n",
    "        os.mkdir(current_output_path)\n",
    "    np.savetxt(os.path.join(current_output_path, \"complete_cell_barcode_count.csv\"), cell_by_gene_complete.astype(np.int), delimiter=',', fmt=\"%d\")\n",
    "    meta_complete = meta_complete.reset_index(drop=True)\n",
    "    meta_complete = meta_complete.rename(columns={\"Unnamed: 0\": \"orig_index\"})\n",
    "    meta_complete.to_csv(os.path.join(current_output_path, \"complete_meta.csv\"))\n",
    "    # cell_barcode_names.to_csv(os.path.join(current_output_path, \"cell_barcode_names.csv\"), header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "0774c9394a1aea0cca30f13f7f373026b44b800bfe001e3b93c40d9a34a50f91"
  },
  "kernelspec": {
   "display_name": "Python 3.8.11 ('connie')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
