{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ce4e8f3c-b9e6-43bb-afd3-fccaf2045643",
   "metadata": {},
   "source": [
    "# Gene cluster visualization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c4969dd-e4a6-4b98-bd8d-34365794360e",
   "metadata": {},
   "source": [
    "2023-04-04"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81ee91d5-f571-4cc2-9dc3-00ed38180fed",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import scanpy as sc\n",
    "import random as rd\n",
    "import matplotlib.pyplot as plt\n",
    "from starmap.sequencing import *\n",
    "from natsort import natsorted\n",
    "from scipy.io import loadmat, savemat\n",
    "from skimage.filters import threshold_otsu\n",
    "from skimage.color import label2rgb\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "from skimage.measure import find_contours\n",
    "from matplotlib.colors import ListedColormap\n",
    "from skimage.segmentation import find_boundaries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef39bd5f-65d0-4680-a485-0856e613cce0",
   "metadata": {},
   "source": [
    "### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26ca5372-2972-40b0-9702-0f0b68a05d7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import wraps\n",
    "from time import time\n",
    "\n",
    "# Timer\n",
    "def timer(func):\n",
    "    @wraps(func)\n",
    "    def _time_it(*args, **kwargs):\n",
    "        start = int(round(time() * 1000))\n",
    "        try:\n",
    "            return func(*args, **kwargs)\n",
    "        finally:\n",
    "            end_ = int(round(time() * 1000)) - start\n",
    "            end_ = round(end_ / 1000, 4)\n",
    "            print(f\"Total execution time: {end_ if end_ > 0 else 0} s\")\n",
    "    return _time_it\n",
    "\n",
    "\n",
    "# Load reads and their positions from mat file\n",
    "def load_reads(fpath, reads_file):\n",
    "    S = loadmat(os.path.join(fpath, reads_file))\n",
    "    bases = [str(i[0][0]) for i in S[\"merged_reads\"]]\n",
    "    points = S[\"merged_points\"]\n",
    "    temp = np.zeros(points.shape)\n",
    "    temp[:, 0] = np.round(points[:, 1]-1)\n",
    "    temp[:, 1] = np.round(points[:, 0]-1)\n",
    "    temp[:, 2] = np.round(points[:, 2]-1)\n",
    "    \n",
    "    print(f\"Number of reads: {len(bases)}\")\n",
    "    \n",
    "    return bases, temp\n",
    "\n",
    "\n",
    "# Load gene table from genes.csv\n",
    "def load_genes(fpath):\n",
    "    genes2seq = {}\n",
    "    seq2genes = {}\n",
    "    with open(os.path.join(fpath, \"genes.csv\"), encoding='utf-8-sig') as f:\n",
    "        for l in f:\n",
    "            fields = l.rstrip().split(\",\")\n",
    "            genes2seq[fields[0]] = \"\".join([str(s+1) for s in encode_SOLID(fields[1][::-1])])\n",
    "            seq2genes[genes2seq[fields[0]]] = fields[0]\n",
    "    return genes2seq, seq2genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb21a222-41ce-40fe-887f-2baecc1a99fd",
   "metadata": {},
   "source": [
    "## Input path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8d9ecef-191f-4c06-b096-b3dad1419a3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# IO path \n",
    "# base_path = 'path/to/2022-03-28-TEMPOmap-images'\n",
    "base_path = 'K:/2022-03-28-TEMPOmap-images/'\n",
    "\n",
    "out_path = os.path.join(base_path, 'output')\n",
    "if not os.path.exists(out_path):\n",
    "    os.mkdir(out_path)\n",
    "    \n",
    "fig_path = os.path.join(base_path, 'figures')\n",
    "if not os.path.exists(fig_path):\n",
    "    os.mkdir(fig_path)\n",
    "    \n",
    "sample_dict = {'20h_labeling': '20h labeling',\n",
    "               '1h_labeling_6h_wash': '1h labeling, 6h wash',\n",
    "               '1h_labeling_4h_wash': '1h labeling, 4h wash',\n",
    "               '1h_labeling_2h_wash': '1h labeling, 2h wash',\n",
    "               '1h_labeling_1h_wash': '1h labeling, 1h wash',\n",
    "               '1h_labeling': '1h labeling'}\n",
    "\n",
    "sample_dirs = list(sample_dict.keys())\n",
    "sample_dirs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7061daf0-5d44-4b61-8fa6-2619c559b70f",
   "metadata": {},
   "source": [
    "## Load gene cluster label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0bcf87e-7962-4116-b53a-ff3782e0ba51",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load m6A \n",
    "m6a_dc1_df = pd.read_excel(os.path.join(base_path, 'gene_modules/2022-01-09-m6A-gene-label.xlsx'))\n",
    "m6a_dc1_df.columns = ['gene', 'm6A label']\n",
    "\n",
    "# modify\n",
    "m6a_dc1_df['m6A label new'] = 999\n",
    "m6a_dc1_df.loc[m6a_dc1_df['m6A label'] == 'm6A', 'm6A label new'] = 0\n",
    "m6a_dc1_df.loc[m6a_dc1_df['m6A label'] == 'non m6A', 'm6A label new'] = 1\n",
    "\n",
    "m6a_dc1_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4aeb43de-9cba-4d18-9765-bce44bdd3408",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load new labels \n",
    "gene_cluster_df = pd.read_excel(os.path.join(base_path, 'gene_modules/2022-05-16-TEMPOmap_new_gene_group.xlsx'))\n",
    "gene_cluster_df.columns = ['gene', 'cluster']\n",
    "gene_cluster_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5061847-19f2-4945-8ed6-be05f1efc0f4",
   "metadata": {},
   "source": [
    "## Visualization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5157489a-d8d7-464c-82cf-8e2e0f59bd5b",
   "metadata": {},
   "source": [
    "### Single cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c7832ae-4c1e-4973-8d2d-da41e5d90586",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot_single_cell_pattern\n",
    "def plot_single_cell_pattern(base_path=None, sample=None, orig_index=None, \n",
    "                             input_df=None, plot_field=None, offset=None, bnd_linewidth=2, spot_size=15, \n",
    "                            nuclei_color=None, cell_color=None, spot_colors=None,\n",
    "                            show=True, save=False, save_path=None, file_name=None):\n",
    "    \n",
    "    # set path \n",
    "    input_path = os.path.join(base_path, sample)\n",
    "\n",
    "    # load image\n",
    "    current_dapi_label = load_label_image(os.path.join(input_path, '2D'), fname='dapi_label.tiff')\n",
    "    current_cell_label = load_label_image(os.path.join(input_path, '2D'), fname='overlay_label.tiff')\n",
    "    current_segmentation = load_label_image(os.path.join(input_path, '3D'), fname='cell.tif')\n",
    "    \n",
    "    # load gene info\n",
    "    genes2seqs, seqs2genes = load_genes(base_path)\n",
    "\n",
    "    # Load reads \n",
    "    bases, points = load_reads(input_path, \"merged_goodPoints_max3d_new.mat\")\n",
    "    bases = np.array(bases)\n",
    "    points = np.array(points)\n",
    "    \n",
    "    # find cell\n",
    "    cell = current_cell_label == orig_index + 1\n",
    "    dapi = np.where(cell, current_dapi_label, 0)\n",
    "\n",
    "    cell_region = None\n",
    "    # print(current_cell_label.max())\n",
    "    for i, region in enumerate(regionprops(current_cell_label)):\n",
    "        label_num = orig_index + 1\n",
    "        if region.label == label_num:\n",
    "            cell_region = region\n",
    "\n",
    "    cell_bbox = cell_region.bbox\n",
    "    cell_crop = cell[cell_bbox[0]:cell_bbox[2], cell_bbox[1]:cell_bbox[3]]\n",
    "    dapi_crop = dapi[cell_bbox[0]:cell_bbox[2], cell_bbox[1]:cell_bbox[3]]\n",
    "    dapi_crop = dapi_crop > 0\n",
    "\n",
    "    # get alignment \n",
    "    bg = np.zeros([offset, offset], np.uint8)\n",
    "    bbox_center = np.array(dapi_crop.shape) / 2\n",
    "    bbox_new_origin = [int(offset / 2 - bbox_center[0]), int(offset / 2 - bbox_center[1])]\n",
    "\n",
    "    # get nuclei / cell contour\n",
    "    bg[bbox_new_origin[0]:bbox_new_origin[0] + cell_crop.shape[0], bbox_new_origin[1]:bbox_new_origin[1] + cell_crop.shape[1]] = cell_crop\n",
    "    cell_bnd = find_contours(bg>0, level=.5)\n",
    "\n",
    "    bg[bbox_new_origin[0]:bbox_new_origin[0] + cell_crop.shape[0], bbox_new_origin[1]:bbox_new_origin[1] + cell_crop.shape[1]] = dapi_crop\n",
    "    dapi_bnd = find_contours(bg>0, level=.5)\n",
    "\n",
    "    # get reads \n",
    "    points = points.astype(int)\n",
    "    reads_assignment = current_segmentation[points[:, 2], points[:, 0], points[:, 1]]\n",
    "\n",
    "    reads_info = pd.DataFrame({'x':points[:, 0], 'y':points[:, 1], 'z':points[:, 2], 'cell_label':reads_assignment})\n",
    "    reads_info = reads_info.astype(np.int32)\n",
    "    reads_info['orig_index'] = reads_info['cell_label'] - 1\n",
    "    reads_info['gene'] = bases\n",
    "    reads_info = reads_info.loc[reads_info['orig_index'] == orig_index, :]\n",
    "    reads_info = pd.merge(reads_info, input_df)\n",
    "\n",
    "    # get reads shift \n",
    "    shifts = [bbox_new_origin[0] - cell_bbox[0], bbox_new_origin[1] - cell_bbox[1]]\n",
    "    reads_info['x'] = reads_info['x'] + shifts[0] \n",
    "    reads_info['y'] = reads_info['y'] + shifts[1] \n",
    "    \n",
    "    # set color\n",
    "    cpl = sns.color_palette(spot_colors)\n",
    "    \n",
    "    bg = np.ones([offset, offset], np.uint8) * 255\n",
    "    # bg = np.zeros([offset, offset], np.uint8)\n",
    "    \n",
    "    plt.subplots(figsize=(10, 10))\n",
    "    plt.imshow(bg, cmap='Greys')\n",
    "    for contour in dapi_bnd:\n",
    "        plt.plot(contour[:, 1], contour[:, 0], linewidth=bnd_linewidth, c=nuclei_color)\n",
    "\n",
    "    for contour in cell_bnd:\n",
    "        plt.plot(contour[:, 1], contour[:, 0], linewidth=bnd_linewidth, c=cell_color)\n",
    "\n",
    "    sns.scatterplot(x=\"y\", y=\"x\", hue=plot_field,\n",
    "                    linewidth=0, s=spot_size, palette=cpl,\n",
    "                    data=reads_info)\n",
    "    plt.axis('off')\n",
    "    \n",
    "    if save:\n",
    "        plt.savefig(os.path.join(save_path, file_name))\n",
    "\n",
    "    if show:\n",
    "        plt.show()\n",
    "    else:\n",
    "        plt.clf()\n",
    "        plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc2fad16-0933-4cf0-a16f-fbe78f59902a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# basic parameters\n",
    "sample = '1h_labeling'\n",
    "orig_index = 2171\n",
    "input_df = gene_cluster_df\n",
    "plot_field = 'cluster'\n",
    "offset = 750\n",
    "# nuclei_color = np.array((255, 255, 255)) / 255\n",
    "# cell_color = np.array((255, 255, 255)) / 255\n",
    "\n",
    "nuclei_color = np.array((0, 0, 0))\n",
    "cell_color = np.array((0, 0, 0))\n",
    "\n",
    "# spot_colors = ['#A895E6', '#22E6A2', '#E6350B', '#EBDB34'] \n",
    "# spot_colors = ['#F8766D', '#7CAE02', '#00BFC4', '#C77CFF'] # original\n",
    "# spot_colors = ['#E6350B', '#32b30e', '#00BFC4', '#ad40ff'] # bright\n",
    "\n",
    "spot_colors = ['#A895E6', '#22E6A2', '#E6350B', '#EBDB34', '#d340e3'] # original 5 colors\n",
    "    \n",
    "file_name = f'white_bg_{sample}_{orig_index}_original.pdf'\n",
    "\n",
    "plot_single_cell_pattern(base_path=base_path, \n",
    "                         sample=sample, \n",
    "                         orig_index=orig_index, \n",
    "                         input_df=input_df, \n",
    "                         plot_field=plot_field, \n",
    "                         offset=offset,\n",
    "                         bnd_linewidth=2, \n",
    "                         spot_size=20, \n",
    "                         nuclei_color=nuclei_color, \n",
    "                         cell_color=cell_color, \n",
    "                         spot_colors=spot_colors,\n",
    "                         show=False, \n",
    "                         save=True, \n",
    "                         save_path=fig_path, \n",
    "                         file_name=file_name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f91da23e-cf0f-4537-a467-2e5cbb4ce9ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# basic parameters\n",
    "cell_dict = {\n",
    "    '20h_labeling': [55, 1262, 963, 222],\n",
    "    '1h_labeling': [2169, 2178, 1098, 1461],\n",
    "    '1h_labeling_2h_wash': [1347, 1369, 1669, 1273],\n",
    "    '1h_labeling_4h_wash': [1694, 1917, 1755, 700],\n",
    "    '1h_labeling_6h_wash': [231, 120, 1600, 1130],\n",
    "}\n",
    "# sample = '20h_labeling'\n",
    "# orig_index = 1263\n",
    "input_df = gene_cluster_df\n",
    "plot_field = 'cluster'\n",
    "offset = 750\n",
    "# nuclei_color = np.array((255, 255, 255)) / 255\n",
    "# cell_color = np.array((255, 255, 255)) / 255\n",
    "\n",
    "nuclei_color = np.array((0, 0, 0))\n",
    "cell_color = np.array((0, 0, 0))\n",
    "\n",
    "# spot_colors = ['#A895E6', '#22E6A2', '#E6350B', '#EBDB34'] \n",
    "# spot_colors = ['#F8766D', '#7CAE02', '#00BFC4', '#C77CFF'] # original\n",
    "# spot_colors = ['#E6350B', '#32b30e', '#00BFC4', '#ad40ff'] # bright\n",
    "\n",
    "spot_colors = ['#F8766D', '#A3A500', '#00BF7D', '#00B0F6', '#E76BF3'] # original 5 colors\n",
    "\n",
    "save_path = os.path.join(fig_path, 'cluster-sc')\n",
    "if not os.path.exists(save_path):\n",
    "    os.mkdir(save_path)\n",
    "\n",
    "for current_sample in tqdm(cell_dict.keys()):\n",
    "    for current_cell in cell_dict[current_sample]:\n",
    "        file_name = f'white_bg_{current_sample}_{current_cell}_original.pdf'\n",
    "        plot_single_cell_pattern(base_path=base_path, \n",
    "                                 sample=current_sample, \n",
    "                                 orig_index=current_cell, \n",
    "                                 input_df=input_df, \n",
    "                                 plot_field=plot_field, \n",
    "                                 offset=offset,\n",
    "                                 bnd_linewidth=2, \n",
    "                                 spot_size=20, \n",
    "                                 nuclei_color=nuclei_color, \n",
    "                                 cell_color=cell_color, \n",
    "                                 spot_colors=spot_colors,\n",
    "                                 show=False, \n",
    "                                 save=True, \n",
    "                                 save_path=save_path, \n",
    "                                 file_name=file_name)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a36fdaac-938f-4854-9405-e8a83a52e885",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Batch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a78e16e0-c07c-4df0-8f0b-9f977b2b7ea4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot_single_cell_pattern\n",
    "def plot_single_cell_pattern_batch(base_path=None, sample=None, orig_index_list=None, \n",
    "                             input_df=None, plot_field=None, offset=None, bnd_linewidth=2, spot_size=15, \n",
    "                            nuclei_color=None, cell_color=None, spot_colors=None,\n",
    "                            show=True, save=False, save_path=None):\n",
    "    \n",
    "    # set path \n",
    "    input_path = os.path.join(base_path, sample)\n",
    "\n",
    "    # load image\n",
    "    current_dapi_label = load_label_image(os.path.join(input_path, '2D'), fname='dapi_label.tiff')\n",
    "    current_cell_label = load_label_image(os.path.join(input_path, '2D'), fname='overlay_label.tiff')\n",
    "    current_segmentation = load_label_image(os.path.join(input_path, '3D'), fname='cell.tif')\n",
    "    \n",
    "    # load gene info\n",
    "    genes2seqs, seqs2genes = load_genes(base_path)\n",
    "\n",
    "    # Load reads \n",
    "    bases, points = load_reads(input_path, \"merged_goodPoints_max3d_new.mat\")\n",
    "    bases = np.array(bases)\n",
    "    points = np.array(points)\n",
    "    \n",
    "    for orig_index in tqdm(orig_index_list):\n",
    "        # find cell\n",
    "        cell = current_cell_label == orig_index + 1\n",
    "        dapi = np.where(cell, current_dapi_label, 0)\n",
    "\n",
    "        cell_region = None\n",
    "        # print(current_cell_label.max())\n",
    "        for i, region in enumerate(regionprops(current_cell_label)):\n",
    "            label_num = orig_index + 1\n",
    "            if region.label == label_num:\n",
    "                cell_region = region\n",
    "\n",
    "        cell_bbox = cell_region.bbox\n",
    "        cell_crop = cell[cell_bbox[0]:cell_bbox[2], cell_bbox[1]:cell_bbox[3]]\n",
    "        dapi_crop = dapi[cell_bbox[0]:cell_bbox[2], cell_bbox[1]:cell_bbox[3]]\n",
    "        dapi_crop = dapi_crop > 0\n",
    "\n",
    "        # get alignment \n",
    "        bg = np.zeros([offset, offset], np.uint8)\n",
    "        bbox_center = np.array(dapi_crop.shape) / 2\n",
    "        bbox_new_origin = [int(offset / 2 - bbox_center[0]), int(offset / 2 - bbox_center[1])]\n",
    "\n",
    "        # get nuclei / cell contour\n",
    "        bg[bbox_new_origin[0]:bbox_new_origin[0] + cell_crop.shape[0], bbox_new_origin[1]:bbox_new_origin[1] + cell_crop.shape[1]] = cell_crop\n",
    "        cell_bnd = find_contours(bg>0, level=.5)\n",
    "\n",
    "        bg[bbox_new_origin[0]:bbox_new_origin[0] + cell_crop.shape[0], bbox_new_origin[1]:bbox_new_origin[1] + cell_crop.shape[1]] = dapi_crop\n",
    "        dapi_bnd = find_contours(bg>0, level=.5)\n",
    "\n",
    "        # get reads \n",
    "        points = points.astype(int)\n",
    "        reads_assignment = current_segmentation[points[:, 2], points[:, 0], points[:, 1]]\n",
    "\n",
    "        reads_info = pd.DataFrame({'x':points[:, 0], 'y':points[:, 1], 'z':points[:, 2], 'cell_label':reads_assignment})\n",
    "        reads_info = reads_info.astype(np.int32)\n",
    "        reads_info['orig_index'] = reads_info['cell_label'] - 1\n",
    "        reads_info['gene'] = bases\n",
    "        reads_info = reads_info.loc[reads_info['orig_index'] == orig_index, :]\n",
    "        reads_info = pd.merge(reads_info, input_df)\n",
    "\n",
    "        # get reads shift \n",
    "        shifts = [bbox_new_origin[0] - cell_bbox[0], bbox_new_origin[1] - cell_bbox[1]]\n",
    "        reads_info['x'] = reads_info['x'] + shifts[0] \n",
    "        reads_info['y'] = reads_info['y'] + shifts[1] \n",
    "\n",
    "        # set color\n",
    "        cpl = sns.color_palette(spot_colors)\n",
    "\n",
    "        bg = np.zeros([offset, offset], np.uint8)\n",
    "        plt.subplots(figsize=(10, 10))\n",
    "        plt.imshow(bg, cmap='Greys')\n",
    "        for contour in dapi_bnd:\n",
    "            plt.plot(contour[:, 1], contour[:, 0], linewidth=bnd_linewidth, c=nuclei_color)\n",
    "\n",
    "        for contour in cell_bnd:\n",
    "            plt.plot(contour[:, 1], contour[:, 0], linewidth=bnd_linewidth, c=cell_color)\n",
    "\n",
    "        sns.scatterplot(x=\"y\", y=\"x\", hue=plot_field,\n",
    "                        linewidth=0, s=spot_size, palette=cpl,\n",
    "                        data=reads_info)\n",
    "        plt.axis('off')\n",
    "\n",
    "        if save:\n",
    "            plt.savefig(os.path.join(save_path, f\"Cell-{orig_index}.pdf\"))\n",
    "\n",
    "        if show:\n",
    "            plt.show()\n",
    "        else:\n",
    "            plt.clf()\n",
    "            plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f743d37-0944-4ba6-9bd9-60e5ed6a7ab0",
   "metadata": {},
   "source": [
    "#### Output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "949552e2-1735-4bd7-aadf-58f1a6c869b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import adata (cell cycle)\n",
    "adata = sc.read_h5ad(os.path.join(base_path, 'output', '2023-04-05-Rena-EU-starmap-cc.h5ad'))\n",
    "\n",
    "# basic parameters\n",
    "input_df = gene_cluster_df\n",
    "plot_field = 'cluster'\n",
    "offset = 750\n",
    "nuclei_color = np.array((0, 0, 0))\n",
    "cell_color = np.array((0, 0, 0))\n",
    "spot_colors = ['#F8766D', '#A3A500', '#00BF7D', '#00B0F6', '#E76BF3'] # original 5 colors\n",
    "cell_to_select = 50\n",
    "\n",
    "# set output path \n",
    "out_path = os.path.join(fig_path, '2022-05-22-cluster-sc')\n",
    "if not os.path.exists(out_path):\n",
    "    os.mkdir(out_path)\n",
    "    \n",
    "# iterate sample \n",
    "for i, sample in enumerate(tqdm(sample_dirs)):\n",
    "    print(f\"Plotting: {sample}\")\n",
    "    \n",
    "    # get orig_index_list\n",
    "    orig_index_list = adata.obs.loc[(adata.obs['sample'] == sample) & (adata.obs['phase_ref'] == 'G1'), 'orig_index'].to_list()\n",
    "    # random sample \n",
    "    orig_index_list = rd.sample(orig_index_list, cell_to_select)\n",
    "\n",
    "    save_path = os.path.join(out_path, sample)\n",
    "    if not os.path.exists(save_path):\n",
    "        os.mkdir(save_path)\n",
    "    \n",
    "    plot_single_cell_pattern_batch(base_path=base_path, \n",
    "                             sample=sample, \n",
    "                             orig_index_list=orig_index_list, \n",
    "                             input_df=input_df, \n",
    "                             plot_field=plot_field, \n",
    "                             offset=offset,\n",
    "                             bnd_linewidth=2, \n",
    "                             spot_size=20, \n",
    "                             nuclei_color=nuclei_color, \n",
    "                             cell_color=cell_color, \n",
    "                             spot_colors=spot_colors,\n",
    "                             show=False, \n",
    "                             save=True, \n",
    "                             save_path=save_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31acfaa9-82f3-470f-90bb-38eb1892cce3",
   "metadata": {},
   "source": [
    "### Sample-wise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "570d0473-1aa7-4b77-937c-9b6c0939cb3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Iterate through each sample dir\n",
    "from datetime import datetime\n",
    "date = datetime.today().strftime('%Y-%m-%d')\n",
    "\n",
    "color_dict = {1:'#A895E6', 2:'#22E6A2', 3:'#E6350B', 4:'#EBDB34', 5:'#d340e3'}\n",
    "plot_field = 'cluster'\n",
    "\n",
    "linewidth = .1\n",
    "\n",
    "for current_dir in sample_dirs:\n",
    "\n",
    "    print(f\"Current sample: {current_dir}\")\n",
    "    input_path = os.path.join(base_path, current_dir)\n",
    "\n",
    "    # Load genes\n",
    "    genes2seqs, seqs2genes = load_genes(base_path)\n",
    "\n",
    "    # Load reads \n",
    "    bases, points = load_reads(input_path, \"merged_goodPoints_max3d_new.mat\")\n",
    "    bases = np.array(bases)\n",
    "    points = np.array(points)\n",
    "    \n",
    "    structure_dict = {'whole_cell': 'cell.tif'}\n",
    "\n",
    "    fig_out_path = base_path + 'figures' + '/' + f'{date}_{plot_field}' + '/' + current_dir\n",
    "    if not os.path.exists(fig_out_path):\n",
    "        os.makedirs(fig_out_path)\n",
    "\n",
    "    # Load raw image\n",
    "    current_dapi_label = load_label_image(os.path.join(input_path, '2D'), fname='dapi_label.tiff')\n",
    "    current_cell_label = load_label_image(os.path.join(input_path, '2D'), fname='overlay_label.tiff')\n",
    "    \n",
    "    cell_bnd = find_boundaries(current_cell_label)\n",
    "    dapi_bnd = find_boundaries(current_dapi_label)\n",
    "\n",
    "    cell_bnd = cell_bnd.astype(np.uint8)\n",
    "    cell_bnd[dapi_bnd] = 2\n",
    "    bg = label2rgb(cell_bnd, colors=[(0,0,0), (0,0,255)], bg_label=0, bg_color=(1,1,1))\n",
    "    \n",
    "    for current_structure in structure_dict.keys():\n",
    "        print(f\"====Processing: {current_structure}====\")\n",
    "\n",
    "        # Load segmentation\n",
    "        current_seg = load_label_image(os.path.join(input_path, '3D'), fname=structure_dict[current_structure])\n",
    "\n",
    "        points = points.astype(int)\n",
    "        reads_assignment = current_seg[points[:, 2], points[:, 0], points[:, 1]]\n",
    "\n",
    "        reads_info = pd.DataFrame({'x':points[:, 0], 'y':points[:, 1], 'z':points[:, 2], 'cell_label':reads_assignment})\n",
    "        reads_info = reads_info.astype(np.int32)\n",
    "        reads_info['orig_index'] = reads_info['cell_label'] - 1\n",
    "        reads_info['gene'] = bases\n",
    "        \n",
    "    for current_cluster in sorted(gene_cluster_df[plot_field].unique()):\n",
    "        if current_cluster == 999:\n",
    "            continue\n",
    "            \n",
    "        print(f\"Current field: {plot_field}, Current cluster: {current_cluster}\")\n",
    "        current_gene_list = gene_cluster_df.loc[gene_cluster_df[plot_field] == current_cluster, 'gene'].to_list()\n",
    "        current_reads_df = reads_info.loc[reads_info['gene'].isin(current_gene_list), :]\n",
    "\n",
    "        # remove unassigned reads\n",
    "        current_reads_df = current_reads_df.loc[current_reads_df['cell_label'] != 0, :]\n",
    "\n",
    "        plt.figure(figsize=(bg.shape[0]/1000, bg.shape[1]/1000), dpi=1000)\n",
    "        plt.imshow(bg)\n",
    "        plt.plot(current_reads_df['y'], current_reads_df['x'], '.', color=color_dict[current_cluster], markersize=.3, markeredgewidth=0.0)\n",
    "        plt.axis('off')\n",
    "        plt.tight_layout(pad=0)\n",
    "        # plt.show()\n",
    "        current_fig_path = f\"{fig_out_path}/cluster_{current_cluster}.tiff\"\n",
    "        plt.savefig(current_fig_path, dpi=1000, bbox_inches='tight', pad_inches=0)\n",
    "        plt.clf()\n",
    "        plt.close()\n",
    "        \n",
    "    plt.figure(figsize=(bg.shape[0]/1000, bg.shape[1]/1000), dpi=1000)\n",
    "    plt.imshow(bg)\n",
    "    \n",
    "    for current_cluster in sorted(gene_cluster_df[plot_field].unique()):\n",
    "        if current_cluster == 999:\n",
    "            continue\n",
    "            \n",
    "        current_gene_list = gene_cluster_df.loc[gene_cluster_df[plot_field] == current_cluster, 'gene'].to_list()\n",
    "        current_reads_df = reads_info.loc[reads_info['gene'].isin(current_gene_list), :]\n",
    "\n",
    "        # remove unassigned reads\n",
    "        current_reads_df = current_reads_df.loc[current_reads_df['cell_label'] != 0, :]\n",
    "\n",
    "        plt.plot(current_reads_df['y'], current_reads_df['x'], '.', color=color_dict[current_cluster], markersize=.3, markeredgewidth=0.0)\n",
    "    plt.axis('off')\n",
    "    plt.tight_layout(pad=0)\n",
    "    # plt.show()\n",
    "    current_fig_path = f\"{fig_out_path}/cluster_all.tiff\"\n",
    "    plt.savefig(current_fig_path, dpi=1000, bbox_inches='tight', pad_inches=0)\n",
    "    plt.clf()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "deb79a91-b71f-4f84-85d3-7f83d8c4e5b4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
