{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "548ad0a3-cbe2-42d4-989c-3a101f694615",
   "metadata": {},
   "source": [
    "# Reads assignment example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "e0e9b88e-12d3-415d-8721-7a4bc3455762",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from starmap.sequencing import *\n",
    "from natsort import natsorted\n",
    "from scipy.io import loadmat, savemat\n",
    "from skimage.filters import threshold_otsu\n",
    "from skimage.color import label2rgb\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c10bf5b-2553-4f33-ad46-1d1ae2eb0b4e",
   "metadata": {},
   "source": [
    "### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "d4a054bf-1aec-477d-9a41-5ee3cfe235c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import wraps\n",
    "from time import time\n",
    "\n",
    "# Timer\n",
    "def timer(func):\n",
    "    @wraps(func)\n",
    "    def _time_it(*args, **kwargs):\n",
    "        start = int(round(time() * 1000))\n",
    "        try:\n",
    "            return func(*args, **kwargs)\n",
    "        finally:\n",
    "            end_ = int(round(time() * 1000)) - start\n",
    "            end_ = round(end_ / 1000, 4)\n",
    "            print(f\"Total execution time: {end_ if end_ > 0 else 0} s\")\n",
    "    return _time_it\n",
    "\n",
    "\n",
    "@timer\n",
    "# Load reads and their positions from mat file\n",
    "def load_reads(fpath, reads_file):\n",
    "    S = loadmat(os.path.join(fpath, reads_file))\n",
    "    bases = [str(i[0][0]) for i in S[\"merged_reads\"]]\n",
    "    points = S[\"merged_points\"]\n",
    "    temp = np.zeros(points.shape)\n",
    "    temp[:, 0] = np.round(points[:, 1]-1)\n",
    "    temp[:, 1] = np.round(points[:, 0]-1)\n",
    "    temp[:, 2] = np.round(points[:, 2]-1)\n",
    "    \n",
    "    print(f\"Number of reads: {len(bases)}\")\n",
    "    \n",
    "    return bases, temp\n",
    "\n",
    "\n",
    "# Load gene table from genes.csv\n",
    "def load_genes(fpath):\n",
    "    genes2seq = {}\n",
    "    seq2genes = {}\n",
    "    with open(os.path.join(fpath, \"genes.csv\"), encoding='utf-8-sig') as f:\n",
    "        for l in f:\n",
    "            fields = l.rstrip().split(\",\")\n",
    "            genes2seq[fields[0]] = \"\".join([str(s+1) for s in encode_SOLID(fields[1][::-1])])\n",
    "            seq2genes[genes2seq[fields[0]]] = fields[0]\n",
    "    return genes2seq, seq2genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11b496c1-d015-4bd7-9a28-7dc3d7238199",
   "metadata": {},
   "source": [
    "## Input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "ed4aaf5a-8a3b-47e0-84e0-a47c1d464856",
   "metadata": {},
   "outputs": [],
   "source": [
    "# IO path \n",
    "base_path = './'\n",
    "out_path = os.path.join(base_path, 'output')\n",
    "if not os.path.exists(out_path):\n",
    "    os.mkdir(out_path)\n",
    "    \n",
    "sample_dict = {'20h_labeling': '20h labeling',\n",
    "               '1h_labeling_6h_wash': '1h labeling, 6h wash',\n",
    "               '1h_labeling_4h_wash': '1h labeling, 4h wash',\n",
    "               '1h_labeling_2h_wash': '1h labeling, 2h wash',\n",
    "               '1h_labeling_1h_wash': '1h labeling, 1h wash',\n",
    "               '1h_labeling': '1h labeling'}\n",
    "\n",
    "sample_dirs = list(sample_dict.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6088db8-79c8-4e0c-a628-b7a389babc02",
   "metadata": {},
   "source": [
    "## Run pipeline for individual sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "8d5344d2-7fdc-4179-88d6-3cdf97d1409a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current sample: 1h_labeling\n",
      "Number of reads: 5219398\n",
      "Total execution time: 35.842 s\n"
     ]
    }
   ],
   "source": [
    "# Iterate through each sample dir\n",
    "current_dir = sample_dirs[0]\n",
    "# print(test_dir)\n",
    "\n",
    "\n",
    "print(f\"Current sample: {current_dir}\")\n",
    "\n",
    "# Load genes\n",
    "genes2seqs, seqs2genes = load_genes(base_path)\n",
    "\n",
    "# Load reads \n",
    "bases, points = load_reads(current_dir, \"merged_goodPoints_max3d_new.mat\")\n",
    "bases = np.array(bases)\n",
    "points = np.array(points)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "c248d9fc-5e16-4fe5-9726-262bf8d18c96",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "====Processing: whole_cell====\n"
     ]
    }
   ],
   "source": [
    "structure_dict = {'whole_cell': 'cell.tif'}\n",
    "\n",
    "total_cells = None\n",
    "expr_out_path = os.path.join(out_path, current_dir)\n",
    "if not os.path.exists(expr_out_path):\n",
    "    os.mkdir(expr_out_path)\n",
    "        \n",
    "expr_dict = {}\n",
    "for current_structure in structure_dict.keys():\n",
    "    print(f\"====Processing: {current_structure}====\")\n",
    "\n",
    "    # Load segmentation\n",
    "    current_seg = load_label_image(current_dir, fname=structure_dict[current_structure])\n",
    "    \n",
    "    # Reads assignment to current_structure\n",
    "    current_out_path = os.path.join(expr_out_path, current_structure)\n",
    "    if not os.path.exists(current_out_path):\n",
    "        os.mkdir(current_out_path)\n",
    "\n",
    "    points = points.astype(int)\n",
    "    reads_assignment = current_seg[points[:, 2], points[:, 0], points[:, 1]]\n",
    "\n",
    "    reads_info = pd.DataFrame({'x':points[:, 0], 'y':points[:, 1], 'z':points[:, 2], 'cell_label':reads_assignment})\n",
    "    reads_info = reads_info.astype(np.int32)\n",
    "    reads_info['orig_index'] = reads_info['cell_label'] - 1\n",
    "    reads_info['gene'] = bases\n",
    "    \n",
    "    reads_info.to_csv(os.path.join(current_out_path, 'reads-info.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "790ca72c-f0d3-4c57-b809-4606bc7b8cdb",
   "metadata": {},
   "source": [
    "## Generate reads pattern plot "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "d82ac3c4-ab4b-4fbb-b2db-a747714afcf7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>cluster_kinetics</th>\n",
       "      <th>cluster_GO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AARS</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ASXL1</td>\n",
       "      <td>1</td>\n",
       "      <td>DNA-binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COX7B</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CTGF</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DLX4</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>803</th>\n",
       "      <td>ZRANB1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>804</th>\n",
       "      <td>ZSCAN12</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA-binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805</th>\n",
       "      <td>ZSCAN29</td>\n",
       "      <td>2</td>\n",
       "      <td>DNA-binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>806</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>1</td>\n",
       "      <td>DNA-binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>807</th>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>808 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene  cluster_kinetics   cluster_GO\n",
       "0       AARS                 2          NaN\n",
       "1      ASXL1                 1  DNA-binding\n",
       "2      COX7B                 3          NaN\n",
       "3       CTGF                 2          NaN\n",
       "4       DLX4                 2          NaN\n",
       "..       ...               ...          ...\n",
       "803   ZRANB1                 1          NaN\n",
       "804  ZSCAN12                 2  DNA-binding\n",
       "805  ZSCAN29                 2  DNA-binding\n",
       "806     ZXDC                 1  DNA-binding\n",
       "807   ZYG11A                 2          NaN\n",
       "\n",
       "[808 rows x 3 columns]"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load gene cluters\n",
    "gene_cluster_df = pd.read_excel('./gene_modules/2022-01-06-gene-cluster-label.xlsx')\n",
    "gene_cluster_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "a71fefc4-e2d7-4dcf-9bcd-0fc8f8db8366",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_GO_dict = {'DNA-binding':0, 'Cell-cell junction':1, 'RNA-binding':2, 'Cell division':3, np.nan:999}\n",
    "gene_cluster_df['cluster_GO'] = gene_cluster_df['cluster_GO'].map(cluster_GO_dict)\n",
    "gene_cluster_df['cluster_GO'] = gene_cluster_df['cluster_GO'].astype(np.int64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "ee146648-8ede-43e3-857d-bd108890baac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current sample: 20h_labeling\n",
      "Number of reads: 3083234\n",
      "Total execution time: 21.824 s\n",
      "====Processing: whole_cell====\n",
      "Current field: cluster_GO, Current cluster: 0\n",
      "Current field: cluster_GO, Current cluster: 1\n",
      "Current field: cluster_GO, Current cluster: 2\n",
      "Current field: cluster_GO, Current cluster: 3\n",
      "Current sample: 1h_labeling_6h_wash\n",
      "Number of reads: 3266939\n",
      "Total execution time: 23.327 s\n",
      "====Processing: whole_cell====\n",
      "Current field: cluster_GO, Current cluster: 0\n",
      "Current field: cluster_GO, Current cluster: 1\n",
      "Current field: cluster_GO, Current cluster: 2\n",
      "Current field: cluster_GO, Current cluster: 3\n",
      "Current sample: 1h_labeling_4h_wash\n",
      "Number of reads: 4825256\n",
      "Total execution time: 34.249 s\n",
      "====Processing: whole_cell====\n",
      "Current field: cluster_GO, Current cluster: 0\n",
      "Current field: cluster_GO, Current cluster: 1\n",
      "Current field: cluster_GO, Current cluster: 2\n",
      "Current field: cluster_GO, Current cluster: 3\n",
      "Current sample: 1h_labeling_2h_wash\n",
      "Number of reads: 4336044\n",
      "Total execution time: 31.909 s\n",
      "====Processing: whole_cell====\n",
      "Current field: cluster_GO, Current cluster: 0\n",
      "Current field: cluster_GO, Current cluster: 1\n",
      "Current field: cluster_GO, Current cluster: 2\n",
      "Current field: cluster_GO, Current cluster: 3\n",
      "Current sample: 1h_labeling_1h_wash\n",
      "Number of reads: 5557806\n",
      "Total execution time: 39.756 s\n",
      "====Processing: whole_cell====\n",
      "Current field: cluster_GO, Current cluster: 0\n",
      "Current field: cluster_GO, Current cluster: 1\n",
      "Current field: cluster_GO, Current cluster: 2\n",
      "Current field: cluster_GO, Current cluster: 3\n",
      "Current sample: 1h_labeling\n",
      "Number of reads: 5219398\n",
      "Total execution time: 37.366 s\n",
      "====Processing: whole_cell====\n",
      "Current field: cluster_GO, Current cluster: 0\n",
      "Current field: cluster_GO, Current cluster: 1\n",
      "Current field: cluster_GO, Current cluster: 2\n",
      "Current field: cluster_GO, Current cluster: 3\n"
     ]
    }
   ],
   "source": [
    "# Iterate through each sample dir\n",
    "test_dirs = sample_dirs[:1]\n",
    "\n",
    "color_dict = {0:'red', 1:'yellow', 2:'cyan', 3:'magenta'}\n",
    "plot_field = 'cluster_GO'\n",
    "\n",
    "for current_dir in sample_dirs:\n",
    "\n",
    "    print(f\"Current sample: {current_dir}\")\n",
    "\n",
    "    # Load genes\n",
    "    genes2seqs, seqs2genes = load_genes(base_path)\n",
    "\n",
    "    # Load reads \n",
    "    bases, points = load_reads(current_dir, \"merged_goodPoints_max3d_new.mat\")\n",
    "    bases = np.array(bases)\n",
    "    points = np.array(points)\n",
    "    \n",
    "    structure_dict = {'whole_cell': 'cell.tif'}\n",
    "\n",
    "    fig_out_path = base_path + 'figures' + '/' + plot_field + '/' + current_dir\n",
    "    if not os.path.exists(fig_out_path):\n",
    "        os.makedirs(fig_out_path)\n",
    "\n",
    "    # Load raw image\n",
    "    current_img = load_label_image(current_dir, fname='cell_bnd.tif')\n",
    "    \n",
    "    for current_structure in structure_dict.keys():\n",
    "        print(f\"====Processing: {current_structure}====\")\n",
    "\n",
    "        # Load segmentation\n",
    "        current_seg = load_label_image(current_dir, fname=structure_dict[current_structure])\n",
    "\n",
    "        points = points.astype(int)\n",
    "        reads_assignment = current_seg[points[:, 2], points[:, 0], points[:, 1]]\n",
    "\n",
    "        reads_info = pd.DataFrame({'x':points[:, 0], 'y':points[:, 1], 'z':points[:, 2], 'cell_label':reads_assignment})\n",
    "        reads_info = reads_info.astype(np.int32)\n",
    "        reads_info['orig_index'] = reads_info['cell_label'] - 1\n",
    "        reads_info['gene'] = bases\n",
    "        \n",
    "    for current_cluster in sorted(gene_cluster_df[plot_field].unique()):\n",
    "        if current_cluster == 999:\n",
    "            continue\n",
    "            \n",
    "        print(f\"Current field: {plot_field}, Current cluster: {current_cluster}\")\n",
    "        current_gene_list = gene_cluster_df.loc[gene_cluster_df[plot_field] == current_cluster, 'gene'].to_list()\n",
    "        current_reads_df = reads_info.loc[reads_info['gene'].isin(current_gene_list), :]\n",
    "\n",
    "        # remove unassigned reads\n",
    "        current_reads_df = current_reads_df.loc[current_reads_df['cell_label'] != 0, :]\n",
    "\n",
    "        plt.figure(figsize=(current_img.shape[0]/1000, current_img.shape[1]/1000), dpi=1000)\n",
    "        plt.imshow(current_img, cmap='gray')\n",
    "        plt.plot(current_reads_df['y'], current_reads_df['x'], '.', color=color_dict[current_cluster], markersize=.3, markeredgewidth=0.0)\n",
    "        plt.axis('off')\n",
    "        plt.tight_layout(pad=0)\n",
    "        # plt.show()\n",
    "        current_fig_path = f\"{fig_out_path}/cluster_{current_cluster}.tiff\"\n",
    "        plt.savefig(current_fig_path, dpi=1000, pil_kwargs={\"compression\": \"tiff_lzw\"}, bbox_inches='tight', pad_inches=0)\n",
    "        plt.clf()\n",
    "        plt.close()\n",
    "        \n",
    "    plt.figure(figsize=(current_img.shape[0]/1000, current_img.shape[1]/1000), dpi=1000)\n",
    "    plt.imshow(current_img, cmap='gray')\n",
    "    for current_cluster in sorted(gene_cluster_df[plot_field].unique()):\n",
    "        if current_cluster == 999:\n",
    "            continue\n",
    "            \n",
    "        current_gene_list = gene_cluster_df.loc[gene_cluster_df[plot_field] == current_cluster, 'gene'].to_list()\n",
    "        current_reads_df = reads_info.loc[reads_info['gene'].isin(current_gene_list), :]\n",
    "\n",
    "        # remove unassigned reads\n",
    "        current_reads_df = current_reads_df.loc[current_reads_df['cell_label'] != 0, :]\n",
    "\n",
    "        plt.plot(current_reads_df['y'], current_reads_df['x'], '.', color=color_dict[current_cluster], markersize=.3, markeredgewidth=0.0)\n",
    "    plt.axis('off')\n",
    "    plt.tight_layout(pad=0)\n",
    "    # plt.show()\n",
    "    current_fig_path = f\"{fig_out_path}/cluster_all.tiff\"\n",
    "    plt.savefig(current_fig_path, dpi=1000, pil_kwargs={\"compression\": \"tiff_lzw\"}, bbox_inches='tight', pad_inches=0)\n",
    "    plt.clf()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c211f224-0e3e-496f-93b2-609513f288cf",
   "metadata": {},
   "source": [
    "## Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "1329df7f-d443-46fa-8f78-195e76c2c681",
   "metadata": {},
   "outputs": [],
   "source": [
    "from skimage.measure import find_contours\n",
    "from matplotlib.colors import ListedColormap\n",
    "from skimage.segmentation import find_boundaries\n",
    "from skimage.color import label2rgb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "f9ce7f10-0569-434e-8a1d-f380bb6ed23a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers).\n"
     ]
    }
   ],
   "source": [
    "\n",
    "cell_bnd = find_boundaries(current_cell_label)\n",
    "dapi_bnd = find_boundaries(current_dapi_label)\n",
    "\n",
    "cell_bnd = cell_bnd.astype(np.uint8)\n",
    "cell_bnd[dapi_bnd] = 2\n",
    "bg = label2rgb(cell_bnd, colors=[(0,0,0), (0,0,255)], bg_label=0, bg_color=(1,1,1))\n",
    "\n",
    "# import matplotlib.pyplot as plt\n",
    "plt.figure(figsize=(current_img.shape[0]/1000, current_img.shape[1]/1000), dpi=1000)\n",
    "plt.imshow(bg)\n",
    "plt.axis('off')\n",
    "plt.tight_layout(pad=0)\n",
    "current_fig_path = f\"./output/test.tiff\"\n",
    "plt.savefig(current_fig_path, dpi=1000, pil_kwargs={\"compression\": \"tiff_lzw\"}, bbox_inches='tight', pad_inches=0)\n",
    "plt.clf()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "6238ac7c-01dc-4f71-a0a2-dce9ab5c1d57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>cluster_kinetics</th>\n",
       "      <th>cluster_GO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AARS</td>\n",
       "      <td>2</td>\n",
       "      <td>999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ASXL1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COX7B</td>\n",
       "      <td>3</td>\n",
       "      <td>999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CTGF</td>\n",
       "      <td>2</td>\n",
       "      <td>999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DLX4</td>\n",
       "      <td>2</td>\n",
       "      <td>999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>803</th>\n",
       "      <td>ZRANB1</td>\n",
       "      <td>1</td>\n",
       "      <td>999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>804</th>\n",
       "      <td>ZSCAN12</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>805</th>\n",
       "      <td>ZSCAN29</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>806</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>807</th>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>2</td>\n",
       "      <td>999</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>808 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene  cluster_kinetics  cluster_GO\n",
       "0       AARS                 2         999\n",
       "1      ASXL1                 1           0\n",
       "2      COX7B                 3         999\n",
       "3       CTGF                 2         999\n",
       "4       DLX4                 2         999\n",
       "..       ...               ...         ...\n",
       "803   ZRANB1                 1         999\n",
       "804  ZSCAN12                 2           0\n",
       "805  ZSCAN29                 2           0\n",
       "806     ZXDC                 1           0\n",
       "807   ZYG11A                 2         999\n",
       "\n",
       "[808 rows x 3 columns]"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_cluster_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "f9779479-0871-4843-a1dc-65bb260cdca2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current sample: 20h_labeling\n",
      "Number of reads: 3083234\n",
      "Total execution time: 22.21 s\n",
      "====Processing: whole_cell====\n",
      "Current field: m6A label new, Current cluster: 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current field: m6A label new, Current cluster: 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers).\n",
      "Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current sample: 1h_labeling_6h_wash\n",
      "Number of reads: 3266939\n",
      "Total execution time: 23.602 s\n",
      "====Processing: whole_cell====\n",
      "Current field: m6A label new, Current cluster: 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current field: m6A label new, Current cluster: 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers).\n",
      "Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current sample: 1h_labeling_4h_wash\n",
      "Number of reads: 4825256\n",
      "Total execution time: 34.858 s\n",
      "====Processing: whole_cell====\n",
      "Current field: m6A label new, Current cluster: 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current field: m6A label new, Current cluster: 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers).\n",
      "Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current sample: 1h_labeling_2h_wash\n",
      "Number of reads: 4336044\n",
      "Total execution time: 36.238 s\n",
      "====Processing: whole_cell====\n",
      "Current field: m6A label new, Current cluster: 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current field: m6A label new, Current cluster: 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers).\n",
      "Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current sample: 1h_labeling_1h_wash\n",
      "Number of reads: 5557806\n",
      "Total execution time: 45.109 s\n",
      "====Processing: whole_cell====\n",
      "Current field: m6A label new, Current cluster: 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current field: m6A label new, Current cluster: 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers).\n",
      "Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current sample: 1h_labeling\n",
      "Number of reads: 5219398\n",
      "Total execution time: 42.251 s\n",
      "====Processing: whole_cell====\n",
      "Current field: m6A label new, Current cluster: 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current field: m6A label new, Current cluster: 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers).\n",
      "Clipping input data to the valid range for imshow with RGB data ([0..1] for floats or [0..255] for integers).\n"
     ]
    }
   ],
   "source": [
    "# Iterate through each sample dir\n",
    "test_dirs = sample_dirs[:1]\n",
    "\n",
    "color_dict = {0:'#F8766D', 1:'#7CAE02', 2:'#00BFC4', 3:'#C77CFF'}\n",
    "# plot_field = 'cluster_kinetics'\n",
    "# plot_field = 'cluster_GO'\n",
    "# plot_field = 'DC1 target'\n",
    "plot_field = 'm6A label new'\n",
    "\n",
    "linewidth = .1\n",
    "\n",
    "for current_dir in sample_dirs:\n",
    "\n",
    "    print(f\"Current sample: {current_dir}\")\n",
    "\n",
    "    # Load genes\n",
    "    genes2seqs, seqs2genes = load_genes(base_path)\n",
    "\n",
    "    # Load reads \n",
    "    bases, points = load_reads(current_dir, \"merged_goodPoints_max3d_new.mat\")\n",
    "    bases = np.array(bases)\n",
    "    points = np.array(points)\n",
    "    \n",
    "    structure_dict = {'whole_cell': 'cell.tif'}\n",
    "\n",
    "    fig_out_path = base_path + 'figures' + '/' + plot_field + '/' + current_dir\n",
    "    if not os.path.exists(fig_out_path):\n",
    "        os.makedirs(fig_out_path)\n",
    "\n",
    "    # Load raw image\n",
    "    current_dapi_label = load_label_image(current_dir, fname='dapi_label.tiff')\n",
    "    current_cell_label = load_label_image(current_dir, fname='overlay_label.tiff')\n",
    "    \n",
    "    cell_bnd = find_boundaries(current_cell_label)\n",
    "    dapi_bnd = find_boundaries(current_dapi_label)\n",
    "\n",
    "    cell_bnd = cell_bnd.astype(np.uint8)\n",
    "    cell_bnd[dapi_bnd] = 2\n",
    "    bg = label2rgb(cell_bnd, colors=[(0,0,0), (0,0,255)], bg_label=0, bg_color=(1,1,1))\n",
    "\n",
    "\n",
    "#     current_dapi_label = current_dapi_label.astype('float')\n",
    "#     dapi_bnd = find_contours(current_dapi_label>0, level=.5)\n",
    "    \n",
    "#     current_cell_label = current_cell_label.astype('float')\n",
    "#     cell_bnd = find_contours(current_cell_label>0, level=.5)\n",
    "\n",
    "#     bg = np.ones(current_dapi_label.shape)\n",
    "    \n",
    "    for current_structure in structure_dict.keys():\n",
    "        print(f\"====Processing: {current_structure}====\")\n",
    "\n",
    "        # Load segmentation\n",
    "        current_seg = load_label_image(current_dir, fname=structure_dict[current_structure])\n",
    "\n",
    "        points = points.astype(int)\n",
    "        reads_assignment = current_seg[points[:, 2], points[:, 0], points[:, 1]]\n",
    "\n",
    "        reads_info = pd.DataFrame({'x':points[:, 0], 'y':points[:, 1], 'z':points[:, 2], 'cell_label':reads_assignment})\n",
    "        reads_info = reads_info.astype(np.int32)\n",
    "        reads_info['orig_index'] = reads_info['cell_label'] - 1\n",
    "        reads_info['gene'] = bases\n",
    "        \n",
    "    for current_cluster in sorted(gene_cluster_df[plot_field].unique()):\n",
    "        if current_cluster == 999:\n",
    "            continue\n",
    "            \n",
    "        print(f\"Current field: {plot_field}, Current cluster: {current_cluster}\")\n",
    "        current_gene_list = gene_cluster_df.loc[gene_cluster_df[plot_field] == current_cluster, 'gene'].to_list()\n",
    "        current_reads_df = reads_info.loc[reads_info['gene'].isin(current_gene_list), :]\n",
    "\n",
    "        # remove unassigned reads\n",
    "        current_reads_df = current_reads_df.loc[current_reads_df['cell_label'] != 0, :]\n",
    "\n",
    "        plt.figure(figsize=(current_img.shape[0]/1000, current_img.shape[1]/1000), dpi=1000)\n",
    "        plt.imshow(bg)\n",
    "#         for contour in dapi_bnd:\n",
    "#             plt.plot(contour[:, 1], contour[:, 0], linewidth=linewidth, c='b')\n",
    "\n",
    "#         for contour in cell_bnd:\n",
    "#             plt.plot(contour[:, 1], contour[:, 0], linewidth=linewidth, c='k')\n",
    "    \n",
    "        plt.plot(current_reads_df['y'], current_reads_df['x'], '.', color=color_dict[current_cluster], markersize=.3, markeredgewidth=0.0)\n",
    "        plt.axis('off')\n",
    "        plt.tight_layout(pad=0)\n",
    "        # plt.show()\n",
    "        current_fig_path = f\"{fig_out_path}/cluster_{current_cluster}.tiff\"\n",
    "        plt.savefig(current_fig_path, dpi=1000, pil_kwargs={\"compression\": \"tiff_lzw\"}, bbox_inches='tight', pad_inches=0)\n",
    "        plt.clf()\n",
    "        plt.close()\n",
    "        \n",
    "    plt.figure(figsize=(current_img.shape[0]/1000, current_img.shape[1]/1000), dpi=1000)\n",
    "    plt.imshow(bg)\n",
    "    \n",
    "#     for contour in dapi_bnd:\n",
    "#         plt.plot(contour[:, 1], contour[:, 0], linewidth=linewidth, c='b')\n",
    "\n",
    "#     for contour in cell_bnd:\n",
    "#         plt.plot(contour[:, 1], contour[:, 0], linewidth=linewidth, c='k')\n",
    "    \n",
    "    for current_cluster in sorted(gene_cluster_df[plot_field].unique()):\n",
    "        if current_cluster == 999:\n",
    "            continue\n",
    "            \n",
    "        current_gene_list = gene_cluster_df.loc[gene_cluster_df[plot_field] == current_cluster, 'gene'].to_list()\n",
    "        current_reads_df = reads_info.loc[reads_info['gene'].isin(current_gene_list), :]\n",
    "\n",
    "        # remove unassigned reads\n",
    "        current_reads_df = current_reads_df.loc[current_reads_df['cell_label'] != 0, :]\n",
    "\n",
    "        plt.plot(current_reads_df['y'], current_reads_df['x'], '.', color=color_dict[current_cluster], markersize=.3, markeredgewidth=0.0)\n",
    "    plt.axis('off')\n",
    "    plt.tight_layout(pad=0)\n",
    "    # plt.show()\n",
    "    current_fig_path = f\"{fig_out_path}/cluster_all.tiff\"\n",
    "    plt.savefig(current_fig_path, dpi=1000, pil_kwargs={\"compression\": \"tiff_lzw\"}, bbox_inches='tight', pad_inches=0)\n",
    "    plt.clf()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "6bf697ec-c40c-4ba3-a561-f296cf00ae0d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>m6A label</th>\n",
       "      <th>DC1 target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DGCR11</td>\n",
       "      <td>other</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SNHG20</td>\n",
       "      <td>other</td>\n",
       "      <td>non target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SSR2</td>\n",
       "      <td>m6A</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TSPAN3</td>\n",
       "      <td>m6A</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AASS</td>\n",
       "      <td>other</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>993</th>\n",
       "      <td>PTRH2</td>\n",
       "      <td>m6A</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>994</th>\n",
       "      <td>CBX4</td>\n",
       "      <td>m6A</td>\n",
       "      <td>target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>NDUFS4</td>\n",
       "      <td>other</td>\n",
       "      <td>non target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>NDUFS7</td>\n",
       "      <td>non m6A</td>\n",
       "      <td>non target</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>HSPA1B</td>\n",
       "      <td>m6A</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>998 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene m6A label  DC1 target\n",
       "0    DGCR11     other       other\n",
       "1    SNHG20     other  non target\n",
       "2      SSR2       m6A       other\n",
       "3    TSPAN3       m6A       other\n",
       "4      AASS     other       other\n",
       "..      ...       ...         ...\n",
       "993   PTRH2       m6A      target\n",
       "994    CBX4       m6A      target\n",
       "995  NDUFS4     other  non target\n",
       "996  NDUFS7   non m6A  non target\n",
       "997  HSPA1B       m6A       other\n",
       "\n",
       "[998 rows x 3 columns]"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load gene cluters\n",
    "gene_cluster_df = pd.read_excel('./gene_modules/2022-01-09-m6A-DC1-gene-label.xlsx')\n",
    "gene_cluster_df.columns = ['gene', 'm6A label', 'DC1 target']\n",
    "gene_cluster_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "9d6051bc-67cf-4080-933c-0f8c9fdafda5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>m6A label</th>\n",
       "      <th>DC1 target</th>\n",
       "      <th>m6A label new</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DGCR11</td>\n",
       "      <td>other</td>\n",
       "      <td>other</td>\n",
       "      <td>999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SNHG20</td>\n",
       "      <td>other</td>\n",
       "      <td>non target</td>\n",
       "      <td>999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SSR2</td>\n",
       "      <td>m6A</td>\n",
       "      <td>other</td>\n",
       "      <td>999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TSPAN3</td>\n",
       "      <td>m6A</td>\n",
       "      <td>other</td>\n",
       "      <td>999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AASS</td>\n",
       "      <td>other</td>\n",
       "      <td>other</td>\n",
       "      <td>999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>993</th>\n",
       "      <td>PTRH2</td>\n",
       "      <td>m6A</td>\n",
       "      <td>target</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>994</th>\n",
       "      <td>CBX4</td>\n",
       "      <td>m6A</td>\n",
       "      <td>target</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>NDUFS4</td>\n",
       "      <td>other</td>\n",
       "      <td>non target</td>\n",
       "      <td>999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>NDUFS7</td>\n",
       "      <td>non m6A</td>\n",
       "      <td>non target</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>HSPA1B</td>\n",
       "      <td>m6A</td>\n",
       "      <td>other</td>\n",
       "      <td>999</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>998 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene m6A label  DC1 target  m6A label new\n",
       "0    DGCR11     other       other            999\n",
       "1    SNHG20     other  non target            999\n",
       "2      SSR2       m6A       other            999\n",
       "3    TSPAN3       m6A       other            999\n",
       "4      AASS     other       other            999\n",
       "..      ...       ...         ...            ...\n",
       "993   PTRH2       m6A      target              0\n",
       "994    CBX4       m6A      target              0\n",
       "995  NDUFS4     other  non target            999\n",
       "996  NDUFS7   non m6A  non target              1\n",
       "997  HSPA1B       m6A       other            999\n",
       "\n",
       "[998 rows x 4 columns]"
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_cluster_df['m6A label new'] = 999\n",
    "gene_cluster_df.loc[(gene_cluster_df['m6A label'] == 'm6A') & (gene_cluster_df['DC1 target'] == 'target'), 'm6A label new'] = 0\n",
    "gene_cluster_df.loc[gene_cluster_df['m6A label'] == 'non m6A', 'm6A label new'] = 1\n",
    "gene_cluster_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "e51a0022-48a9-4490-aea3-88f7736378e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "field = 'm6A label'\n",
    "cluster_GO_dict = {'m6A':0, 'non m6A':1, 'other':2}\n",
    "\n",
    "# field = 'DC1 target'\n",
    "# cluster_GO_dict = {'target':0, 'non target':1, 'other':2}\n",
    "\n",
    "gene_cluster_df[field] = gene_cluster_df[field].map(cluster_GO_dict)\n",
    "gene_cluster_df[field] = gene_cluster_df[field].astype(np.int64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "75fdf10d-228b-4dfe-93b5-8eb39abe91bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current sample: 20h_labeling\n",
      "Number of reads: 3083234\n",
      "Total execution time: 25.328 s\n",
      "====Processing: whole_cell====\n",
      "Current field: DC1 target, Current cluster: 0\n",
      "Current field: DC1 target, Current cluster: 1\n",
      "Current field: DC1 target, Current cluster: 2\n"
     ]
    }
   ],
   "source": [
    "# Iterate through each sample dir\n",
    "test_dirs = sample_dirs[:1]\n",
    "\n",
    "color_dict = {0:'#F8766D', 1:'#7CAE02', 2:'#00BFC4', 3:'#C77CFF'}\n",
    "plot_field = field\n",
    "# plot_field = 'cluster_GO'\n",
    "linewidth = .1\n",
    "\n",
    "for current_dir in test_dirs:\n",
    "\n",
    "    print(f\"Current sample: {current_dir}\")\n",
    "\n",
    "    # Load genes\n",
    "    genes2seqs, seqs2genes = load_genes(base_path)\n",
    "\n",
    "    # Load reads \n",
    "    bases, points = load_reads(current_dir, \"merged_goodPoints_max3d_new.mat\")\n",
    "    bases = np.array(bases)\n",
    "    points = np.array(points)\n",
    "    \n",
    "    structure_dict = {'whole_cell': 'cell.tif'}\n",
    "\n",
    "    fig_out_path = base_path + 'figures' + '/' + plot_field + '/' + current_dir\n",
    "    if not os.path.exists(fig_out_path):\n",
    "        os.makedirs(fig_out_path)\n",
    "\n",
    "    # Load raw image\n",
    "    current_dapi_label = load_label_image(current_dir, fname='dapi_label.tiff')\n",
    "    current_cell_label = load_label_image(current_dir, fname='overlay_label.tiff')\n",
    "    \n",
    "    current_dapi_label = current_dapi_label.astype('float')\n",
    "    dapi_bnd = find_contours(current_dapi_label>0, level=.5)\n",
    "    \n",
    "    current_cell_label = current_cell_label.astype('float')\n",
    "    cell_bnd = find_contours(current_cell_label>0, level=.5)\n",
    "\n",
    "    bg = np.ones(current_dapi_label.shape)\n",
    "    \n",
    "    for current_structure in structure_dict.keys():\n",
    "        print(f\"====Processing: {current_structure}====\")\n",
    "\n",
    "        # Load segmentation\n",
    "        current_seg = load_label_image(current_dir, fname=structure_dict[current_structure])\n",
    "\n",
    "        points = points.astype(int)\n",
    "        reads_assignment = current_seg[points[:, 2], points[:, 0], points[:, 1]]\n",
    "\n",
    "        reads_info = pd.DataFrame({'x':points[:, 0], 'y':points[:, 1], 'z':points[:, 2], 'cell_label':reads_assignment})\n",
    "        reads_info = reads_info.astype(np.int32)\n",
    "        reads_info['orig_index'] = reads_info['cell_label'] - 1\n",
    "        reads_info['gene'] = bases\n",
    "        \n",
    "    for current_cluster in sorted(gene_cluster_df[plot_field].unique()):\n",
    "        if current_cluster == 999:\n",
    "            continue\n",
    "            \n",
    "        print(f\"Current field: {plot_field}, Current cluster: {current_cluster}\")\n",
    "        current_gene_list = gene_cluster_df.loc[gene_cluster_df[plot_field] == current_cluster, 'gene'].to_list()\n",
    "        current_reads_df = reads_info.loc[reads_info['gene'].isin(current_gene_list), :]\n",
    "\n",
    "        # remove unassigned reads\n",
    "        current_reads_df = current_reads_df.loc[current_reads_df['cell_label'] != 0, :]\n",
    "\n",
    "        plt.figure(figsize=(current_img.shape[0]/1000, current_img.shape[1]/1000), dpi=1000)\n",
    "        plt.imshow(bg, cmap='gray_r')\n",
    "        for contour in dapi_bnd:\n",
    "            plt.plot(contour[:, 1], contour[:, 0], linewidth=linewidth, c='b')\n",
    "\n",
    "        for contour in cell_bnd:\n",
    "            plt.plot(contour[:, 1], contour[:, 0], linewidth=linewidth, c='g')\n",
    "    \n",
    "        plt.plot(current_reads_df['y'], current_reads_df['x'], '.', color=color_dict[current_cluster], markersize=.3, markeredgewidth=0.0)\n",
    "        plt.axis('off')\n",
    "        plt.tight_layout(pad=0)\n",
    "        # plt.show()\n",
    "        current_fig_path = f\"{fig_out_path}/cluster_{current_cluster}.tiff\"\n",
    "        plt.savefig(current_fig_path, dpi=1000, pil_kwargs={\"compression\": \"tiff_lzw\"}, bbox_inches='tight', pad_inches=0)\n",
    "        plt.clf()\n",
    "        plt.close()\n",
    "        \n",
    "    plt.figure(figsize=(current_img.shape[0]/1000, current_img.shape[1]/1000), dpi=1000)\n",
    "    plt.imshow(bg, cmap='gray_r')\n",
    "    for contour in dapi_bnd:\n",
    "        plt.plot(contour[:, 1], contour[:, 0], linewidth=linewidth, c='b')\n",
    "\n",
    "    for contour in cell_bnd:\n",
    "        plt.plot(contour[:, 1], contour[:, 0], linewidth=linewidth, c='g')\n",
    "    \n",
    "    for current_cluster in sorted(gene_cluster_df[plot_field].unique()):\n",
    "        if current_cluster == 999:\n",
    "            continue\n",
    "            \n",
    "        current_gene_list = gene_cluster_df.loc[gene_cluster_df[plot_field] == current_cluster, 'gene'].to_list()\n",
    "        current_reads_df = reads_info.loc[reads_info['gene'].isin(current_gene_list), :]\n",
    "\n",
    "        # remove unassigned reads\n",
    "        current_reads_df = current_reads_df.loc[current_reads_df['cell_label'] != 0, :]\n",
    "\n",
    "        plt.plot(current_reads_df['y'], current_reads_df['x'], '.', color=color_dict[current_cluster], markersize=.3, markeredgewidth=0.0)\n",
    "    plt.axis('off')\n",
    "    plt.tight_layout(pad=0)\n",
    "    # plt.show()\n",
    "    current_fig_path = f\"{fig_out_path}/cluster_all.tiff\"\n",
    "    plt.savefig(current_fig_path, dpi=1000, pil_kwargs={\"compression\": \"tiff_lzw\"}, bbox_inches='tight', pad_inches=0)\n",
    "    plt.clf()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dad0d9bc-036f-478d-bbec-4c6443d25621",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
